<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yongjin Park" />

<meta name="date" content="2024-02-07" />

<title>Step 3: differential Expression analysis</title>

<script src="step3_deg_analysis_files/header-attrs-2.25/header-attrs.js"></script>
<script src="step3_deg_analysis_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="step3_deg_analysis_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="step3_deg_analysis_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="step3_deg_analysis_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="step3_deg_analysis_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="step3_deg_analysis_files/navigation-1.1/tabsets.js"></script>
<link href="step3_deg_analysis_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="step3_deg_analysis_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Step 3: differential Expression analysis</h1>
<h4 class="author">Yongjin Park</h4>
<h4 class="date">2024-02-07</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#ms-vs.-hc-at-the-major-cell-type-level">MS vs. HC at the major cell type level</a>
<ul>
<li><a href="#comparison-with-the-bulk-deg-results">Comparison with the bulk DEG results</a></li>
<li><a href="#memory-treg-genes">1. Memory Treg genes</a>
<ul>
<li><a href="#correlation-0.23-with-p-value-5.8e-98">Correlation 0.23 with p-value = 5.8e-98</a></li>
</ul></li>
<li><a href="#memory-tconv-genes">2. Memory Tconv genes</a>
<ul>
<li><a href="#correlation-0.26-with-p-value-1.71e-117">Correlation 0.26 with p-value = 1.71e-117</a></li>
<li><a href="#found-784-unique-genes-strongly-perturbed-by-ms-with-fwer-5">Found 784 unique genes strongly perturbed by MS with FWER 5%</a></li>
</ul></li>
<li><a href="#show-genes-with-fwer-10-2-overlapping-with-the-bulk-degs">3. Show genes with FWER <span class="math inline">\(&lt; 10^{-2}\)</span> overlapping with the bulk DEGs</a></li>
<li><a href="#additional-genesproteins">4. Additional genes/proteins</a></li>
<li><a href="#appendix-prdm1-short-vs.-long">5. Appendix: PRDM1 short vs. long</a></li>
</ul></li>
</ul>
</div>

<pre class="r"><code>.data &lt;- fileset.list(&quot;result/step2/matrix_final&quot;)</code></pre>
<pre class="r"><code>.features &lt;- readLines(.data$row)
.hash &lt;- .features[str_detect(.features, &quot;Hash&quot;)]
.hash.hdr &lt;- &quot;result/step1/hash&quot;
.hash.data &lt;- fileset.list(.hash.hdr)
.hash.info &lt;- read.hash(.hash.data)</code></pre>
<div id="ms-vs.-hc-at-the-major-cell-type-level" class="section level1">
<h1>MS vs. HC at the major cell type level</h1>
<pre class="r"><code>annot.dt &lt;-
    fread(&quot;Tab/step2_celltype.txt.gz&quot;) %&gt;%
    left_join(.hash.info) %&gt;%
    na.omit()</code></pre>
<pre class="r"><code>.mkdir(&quot;result/step3/deg/&quot;)
.deg.data &lt;- fileset.list(&quot;result/step3/deg/hc_ms&quot;)

if.needed(.deg.data, {
    .deg.data &lt;-
        rcpp_mmutil_copy_selected_columns(.data$mtx,
                                          .data$row,
                                          .data$col,
                                          unique(annot.dt$tag),
                                          &quot;result/step3/deg/hc_ms&quot;)
})</code></pre>
<pre class="r"><code>.file &lt;- &quot;result/step3/deg/hc_ms.rds&quot;

if.needed(.file, {

    .celltype &lt;-
        annot.dt[, .(tag, celltype)] %&gt;%
        as.data.frame()

    .cell2indv &lt;- annot.dt[, .(tag, subject)] %&gt;%
        unique %&gt;%
        as.data.frame()

    .indv2exp &lt;- .cell2indv %&gt;%
        select(subject) %&gt;%
        mutate(disease = substr(`subject`, 1, 2)) %&gt;%
        as.data.frame()

    .deg.stat &lt;-
        make.cocoa(.deg.data, .celltype, .cell2indv, .indv2exp,
                   .rank = 15, .em.iter = 20, .em.tol = 1e-8, .take.ln = TRUE,
                   knn = 50, impute.by.knn = TRUE, num.threads = 8)

   saveRDS(.deg.stat, .file)
})

.deg.stat &lt;- readRDS(.file)

.cts &lt;- unique(annot.dt$celltype)
.indvs &lt;- unique(annot.dt$subject)

.hc.ms.dt &lt;-
    list(tot = sort.col(.deg.stat$sum, .cts, .indvs),
         cfa = sort.col(.deg.stat$resid.ln.mu, .cts, .indvs),
         cfa.sd = sort.col(.deg.stat$resid.ln.mu.sd, .cts, .indvs)) %&gt;%
    combine.statistics()

.hc.ms.dt[, c(&quot;subject&quot;,&quot;celltype&quot;) := tstrsplit(`Var2`,split=&quot;_&quot;)]
.hc.ms.dt[, disease := substr(`subject`,1,2)]
.hc.ms.dt[, gene := `Var1`]
.hc.ms.dt[, `Var1` := NULL]
.hc.ms.dt[, `Var2` := NULL]

hc.ms.deg &lt;- summarize.deg(.hc.ms.dt)</code></pre>
<div id="comparison-with-the-bulk-deg-results" class="section level3">
<h3>Comparison with the bulk DEG results</h3>
<pre class="r"><code>read.bulk &lt;- function(.file) {
    fread(.file) %&gt;%
        rename(hgnc_symbol = gene_name) %&gt;%
        rename(bulk.t = t, bulk.pv = P.Value, bulk.qv = adj.P.Val) %&gt;%
        dplyr::select(hgnc_symbol, starts_with(&quot;bulk&quot;))
}

.file &lt;- &quot;data/DEG/20180513/deg.ms.hc.treg.mem.exvivo.sex_covar.ruv.txt&quot;
bulk.mTreg.dt &lt;- read.bulk(.file)

.file &lt;- &quot;data/DEG/20180513/deg.ms.hc.teff.mem.exvivo.sex_covar.ruv.txt&quot;
bulk.mTconv.dt &lt;- read.bulk(.file)</code></pre>
<p><a href="Tab/DEG_MS_vs_HC.txt.gz"><strong>DOWNLOAD:</strong> DEG MS vs HC</a></p>
</div>
<div id="memory-treg-genes" class="section level2">
<h2>1. Memory Treg genes</h2>
<pre class="r"><code>.sc &lt;- hc.ms.deg[celltype == &quot;mTreg&quot; &amp;
                 sign(ADD) == sign(z) &amp;
                 sign(ADC) == sign(z),]

.sc[, c(&quot;ensembl_gene_id&quot;,&quot;hgnc_symbol&quot;) := tstrsplit(`gene`, split=&quot;_&quot;)]

.bulk &lt;- bulk.mTreg.dt

.out &lt;- plot.bulk.sc(.sc,
                     .bulk,
                     fdr.cutoff = .01,
                     qv.cutoff = .2,
                     n.top = 7)</code></pre>
<div id="correlation-0.23-with-p-value-5.8e-98" class="section level3">
<h3>Correlation 0.23 with p-value = 5.8e-98</h3>
<p><img src="Fig/STEP3/Fig_DEG_comparison_mTreg-1.png" /><!-- --></p>
<p><a href="Fig/STEP3//Fig_DEG_comparison_mTreg.pdf">PDF</a></p>
</div>
</div>
<div id="memory-tconv-genes" class="section level2">
<h2>2. Memory Tconv genes</h2>
<pre class="r"><code>.sc &lt;- hc.ms.deg[celltype == &quot;mTconv&quot; &amp;
                 sign(ADD) == sign(z) &amp;
                 sign(ADC) == sign(z),]

.sc[, c(&quot;ensembl_gene_id&quot;,&quot;hgnc_symbol&quot;) := tstrsplit(`gene`, split=&quot;_&quot;)]

.bulk &lt;- bulk.mTconv.dt

.out &lt;- plot.bulk.sc(.sc,
                     .bulk,
                     fdr.cutoff = .01,
                     qv.cutoff = .2,
                     n.top = 5)</code></pre>
<div id="correlation-0.26-with-p-value-1.71e-117" class="section level3">
<h3>Correlation 0.26 with p-value = 1.71e-117</h3>
<p><img src="Fig/STEP3/Fig_DEG_comparison_mTconv-1.png" /><!-- --></p>
<p><a href="Fig/STEP3//Fig_DEG_comparison_mTconv.pdf">PDF</a></p>
</div>
<div id="found-784-unique-genes-strongly-perturbed-by-ms-with-fwer-5" class="section level3">
<h3>Found 784 unique genes strongly perturbed by MS with FWER 5%</h3>
<ul>
<li><p>Up-regulated: 214</p></li>
<li><p>Down-regulated: 608</p></li>
<li><p>Total pairs of genes and cell types: 48,428</p></li>
</ul>
<pre class="r"><code>count.deg &lt;- function(.dt, fwer.cutoff = .05) {
    .dt[fwer &lt; fwer.cutoff &amp;
        sign(ADD) == sign(ADE) &amp;
        sign(ADC) == sign(ADE),
        .(n = .N),
        by = .(celltype,
               direction = if_else(z &gt; 0, &quot;up&quot;, &quot;down&quot;))
        ] %&gt;%
        mutate(direction = factor(direction, c(&quot;up&quot;, &quot;down&quot;))) %&gt;%
        group_by(celltype) %&gt;%
        arrange(desc(direction)) %&gt;%
        mutate(nc = cumsum(n)) %&gt;%
        ungroup
}</code></pre>
<p><img src="Fig/STEP3/Fig_DEG_count-1.png" /><!-- --></p>
<p><a href="Fig/STEP3//Fig_DEG_count.pdf">PDF</a></p>
</div>
</div>
<div id="show-genes-with-fwer-10-2-overlapping-with-the-bulk-degs" class="section level2">
<h2>3. Show genes with FWER <span class="math inline">\(&lt; 10^{-2}\)</span> overlapping with the bulk DEGs</h2>
<pre class="r"><code>show.each.gene &lt;- function(g, .data, .stat, .cutoff = 0) {

    .dt.g &lt;- na.omit(.data[gene == g]) %&gt;% filter(tot &gt; .cutoff)
    .stat.dt &lt;- na.omit(.stat[gene == g]) %&gt;% rename(total = tot)

    .dir &lt;- .stat.dt[which.max(abs(.stat.dt$ADE)), .(ADE)] %&gt;%
        unlist

    if(.dir &gt; 0) {

        .dt &lt;- .dt.g %&gt;%
            left_join(.stat.dt, by = c(&quot;celltype&quot;, &quot;gene&quot;)) %&gt;%
            group_by(disease, celltype) %&gt;%
            arrange(cfa) %&gt;%
            mutate(j = 1:n()) %&gt;%
            ungroup() %&gt;%
            mutate(x = disease %&amp;% &quot;_&quot; %&amp;% j)

    } else {

        .dt &lt;- .dt.g %&gt;%
            left_join(.stat.dt, by = c(&quot;celltype&quot;, &quot;gene&quot;)) %&gt;%
            group_by(disease, celltype) %&gt;%
            arrange(desc(cfa)) %&gt;%
            mutate(j = 1:n()) %&gt;%
            ungroup() %&gt;%
            mutate(x = disease %&amp;% &quot;_&quot; %&amp;% j)

    }

    .dt &lt;- .dt %&gt;%
        mutate(.pv = if_else(is.na(pv), &quot;1&quot;, num.sci(pv))) %&gt;%
        mutate(.facet = celltype %&amp;% &quot;\n(&quot; %&amp;% .pv %&amp;% &quot;)&quot;) %&gt;%
        mutate(ADD.z = ADD/ADD.se, ADC.z = ADC/ADC.se) %&gt;%
        as.data.table

    .stat.dt &lt;- .dt[, .(celltype,
                        ADE, ADE.se,
                        ADD, ADD.se,
                        ADC, ADC.se,
                        pv, .facet)] %&gt;%
        unique

    .aes &lt;- aes(x = x,
                y = cfa,
                ymin = cfa - 2 * cfa.sd,
                ymax = cfa + 2 * cfa.sd,
                fill = disease)

    .scale.y &lt;- scale_y_continuous(labels = function(x) num.round(exp(x)))

    p1 &lt;-
        .gg.plot(.dt, .aes) +
        theme(legend.position = &quot;top&quot;) +
        facet_wrap(~.facet, nrow = 1, scales=&quot;free&quot;) +
        .scale.y +
        theme(axis.ticks.x = element_blank()) +
        theme(axis.text.x = element_blank()) +
        theme(axis.title.x = element_blank()) +
        geom_hline(yintercept = 0, linewidth = 1, colour = &quot;gray&quot;) +
        geom_hline(data = .stat.dt, aes(yintercept = ADE), linewidth = 1, colour = &quot;#F8766D&quot;) +
        geom_linerange(linewidth = .2) +
        geom_point(aes(size = tot), pch = 21, stroke = .2) +
        scale_size_continuous(&quot;total(k)&quot;, range=c(0.1, 2.5), labels = function(x) num.int(x/1e3)) +
        scale_fill_manual(values = c(&quot;gray90&quot;,&quot;#F8766D&quot;)) +
        xlab(&quot;&quot;) + ylab(&quot;fold change&quot;) + ggtitle(g)

    .temp.1 &lt;- .stat.dt %&gt;%
        dplyr::select(celltype, pv, ADE, ADD, ADC) %&gt;%
        mutate(ADC = - ADC) %&gt;%                  # flip the effect
        melt(id.vars=c(&quot;celltype&quot;, &quot;pv&quot;), value.name = &quot;mu&quot;)

    .temp.2 &lt;- .stat.dt %&gt;%
        dplyr::select(celltype, ADE.se, ADD.se, ADC.se) %&gt;%
        melt(id.vars=&quot;celltype&quot;, value.name = &quot;se&quot;) %&gt;%
        mutate(variable = str_remove(variable, &quot;.se&quot;))

    .temp &lt;- left_join(.temp.1, .temp.2, by = c(&quot;celltype&quot;, &quot;variable&quot;))

    .aes &lt;- aes(x = variable,
                y = mu,
                ymin = mu - 3 * se,
                ymax = mu + 3 * se)

    p2 &lt;-
        .gg.plot(.temp, .aes) +
        theme(axis.text.x = element_text(angle=90, hjust = 1, vjust= .5)) +
        facet_wrap(~celltype, nrow = 1, scales=&quot;free&quot;) +
        .scale.y + ylab(&quot;fold change&quot;) + xlab(&quot;&quot;) +
        geom_hline(yintercept = 0, lty = 2, linewidth = .5, colour = &quot;black&quot;) +
        geom_linerange(linewidth = .2) +
        geom_point(aes(fill = variable, stroke = .2, size = -log10(pv)), pch = 21) +
        scale_fill_brewer(palette = &quot;Greens&quot;, guide=&quot;none&quot;) +
        scale_size_continuous(&quot;p-value&quot;, range=c(.5, 2.5), breaks=seq(0, 10, 2), labels = function(x) 10^(-x))

    plt &lt;- (p1 / p2) + patchwork::plot_layout(heights = c(4,2))
    return(plt)
}</code></pre>
<pre class="r"><code>.bulk.genes &lt;-
    rbind(bulk.mTreg.dt[bulk.qv &lt; .2, .(hgnc_symbol)],
          bulk.mTconv.dt[bulk.qv &lt; .2, .(hgnc_symbol)]) %&gt;%
    unique() %&gt;%
    unlist()

hc.ms.deg[, c(&quot;ensembl_gene_id&quot;, &quot;hgnc_symbol&quot;) := tstrsplit(`gene`, split=&quot;_&quot;)]

.genes.show &lt;-
    hc.ms.deg[fwer &lt; 0.05 &amp;
              hgnc_symbol %in% .bulk.genes &amp;
              sign(ADD) == sign(ADE) &amp;
              sign(ADC) == sign(ADE)] %&gt;%
    select(gene) %&gt;%
    unique %&gt;%
    unlist %&gt;%
    as.character</code></pre>
<p><a href="Fig/STEP3//Fig_DEG_example_CD226_CD226.pdf">CD226_CD226</a> <a href="Fig/STEP3//Fig_DEG_example_CD27_CD27.pdf">CD27_CD27</a> <a href="Fig/STEP3//Fig_DEG_example_CD58_CD58.pdf">CD58_CD58</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000003056_M6PR.pdf">ENSG00000003056_M6PR</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000009790_TRAF3IP3.pdf">ENSG00000009790_TRAF3IP3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000009844_VTA1.pdf">ENSG00000009844_VTA1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000023330_ALAS1.pdf">ENSG00000023330_ALAS1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000034677_RNF19A.pdf">ENSG00000034677_RNF19A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000038274_MAT2B.pdf">ENSG00000038274_MAT2B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000040633_PHF23.pdf">ENSG00000040633_PHF23</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000043462_LCP2.pdf">ENSG00000043462_LCP2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000051108_HERPUD1.pdf">ENSG00000051108_HERPUD1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000057657_PRDM1.pdf">ENSG00000057657_PRDM1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000065154_OAT.pdf">ENSG00000065154_OAT</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000065911_MTHFD2.pdf">ENSG00000065911_MTHFD2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000069399_BCL3.pdf">ENSG00000069399_BCL3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000070756_PABPC1.pdf">ENSG00000070756_PABPC1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000071073_MGAT4A.pdf">ENSG00000071073_MGAT4A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000073578_SDHA.pdf">ENSG00000073578_SDHA</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000075239_ACAT1.pdf">ENSG00000075239_ACAT1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000075292_ZNF638.pdf">ENSG00000075292_ZNF638</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000075420_FNDC3B.pdf">ENSG00000075420_FNDC3B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000075426_FOSL2.pdf">ENSG00000075426_FOSL2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000077420_APBB1IP.pdf">ENSG00000077420_APBB1IP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000079246_XRCC5.pdf">ENSG00000079246_XRCC5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000081237_PTPRC.pdf">ENSG00000081237_PTPRC</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000087502_ERGIC2.pdf">ENSG00000087502_ERGIC2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000088986_DYNLL1.pdf">ENSG00000088986_DYNLL1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000089157_RPLP0.pdf">ENSG00000089157_RPLP0</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000092094_OSGEP.pdf">ENSG00000092094_OSGEP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000092621_PHGDH.pdf">ENSG00000092621_PHGDH</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000095370_SH2D3C.pdf">ENSG00000095370_SH2D3C</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000099622_CIRBP.pdf">ENSG00000099622_CIRBP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000099985_OSM.pdf">ENSG00000099985_OSM</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100100_PIK3IP1.pdf">ENSG00000100100_PIK3IP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100129_EIF3L.pdf">ENSG00000100129_EIF3L</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100201_DDX17.pdf">ENSG00000100201_DDX17</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100219_XBP1.pdf">ENSG00000100219_XBP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100316_RPL3.pdf">ENSG00000100316_RPL3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000100650_SRSF5.pdf">ENSG00000100650_SRSF5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000101608_MYL12A.pdf">ENSG00000101608_MYL12A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000102245_CD40LG.pdf">ENSG00000102245_CD40LG</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000102760_RGCC.pdf">ENSG00000102760_RGCC</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000104529_EEF1D.pdf">ENSG00000104529_EEF1D</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000104904_OAZ1.pdf">ENSG00000104904_OAZ1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000106153_CHCHD2.pdf">ENSG00000106153_CHCHD2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000106460_TMEM106B.pdf">ENSG00000106460_TMEM106B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000106560_GIMAP2.pdf">ENSG00000106560_GIMAP2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000107201_DDX58.pdf">ENSG00000107201_DDX58</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000108298_RPL19.pdf">ENSG00000108298_RPL19</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000108622_ICAM2.pdf">ENSG00000108622_ICAM2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000109062_SLC9A3R1.pdf">ENSG00000109062_SLC9A3R1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000109321_AREG.pdf">ENSG00000109321_AREG</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000109861_CTSC.pdf">ENSG00000109861_CTSC</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000109971_HSPA8.pdf">ENSG00000109971_HSPA8</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000110066_KMT5B.pdf">ENSG00000110066_KMT5B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000110852_CLEC2B.pdf">ENSG00000110852_CLEC2B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000111639_MRPL51.pdf">ENSG00000111639_MRPL51</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000111875_ASF1A.pdf">ENSG00000111875_ASF1A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000112079_STK38.pdf">ENSG00000112079_STK38</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000112110_MRPL18.pdf">ENSG00000112110_MRPL18</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000113732_ATP6V0E1.pdf">ENSG00000113732_ATP6V0E1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000114209_PDCD10.pdf">ENSG00000114209_PDCD10</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000114395_CYB561D2.pdf">ENSG00000114395_CYB561D2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000114942_EEF1B2.pdf">ENSG00000114942_EEF1B2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000115091_ACTR3.pdf">ENSG00000115091_ACTR3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000115687_PASK.pdf">ENSG00000115687_PASK</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000115758_ODC1.pdf">ENSG00000115758_ODC1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000116521_SCAMP3.pdf">ENSG00000116521_SCAMP3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000116679_IVNS1ABP.pdf">ENSG00000116679_IVNS1ABP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000116717_GADD45A.pdf">ENSG00000116717_GADD45A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000117090_SLAMF1.pdf">ENSG00000117090_SLAMF1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000117318_ID3.pdf">ENSG00000117318_ID3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000118515_SGK1.pdf">ENSG00000118515_SGK1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000119718_EIF2B2.pdf">ENSG00000119718_EIF2B2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000120875_DUSP4.pdf">ENSG00000120875_DUSP4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000120913_PDLIM2.pdf">ENSG00000120913_PDLIM2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000120963_ZNF706.pdf">ENSG00000120963_ZNF706</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000121742_GJB6.pdf">ENSG00000121742_GJB6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000121858_TNFSF10.pdf">ENSG00000121858_TNFSF10</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000122026_RPL21.pdf">ENSG00000122026_RPL21</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000122224_LY9.pdf">ENSG00000122224_LY9</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000122406_RPL5.pdf">ENSG00000122406_RPL5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000122862_SRGN.pdf">ENSG00000122862_SRGN</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000123416_TUBA1B.pdf">ENSG00000123416_TUBA1B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000124588_NQO2.pdf">ENSG00000124588_NQO2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000125148_MT2A.pdf">ENSG00000125148_MT2A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000125384_PTGER2.pdf">ENSG00000125384_PTGER2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000125898_FAM110A.pdf">ENSG00000125898_FAM110A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000126524_SBDS.pdf">ENSG00000126524_SBDS</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000129562_DAD1.pdf">ENSG00000129562_DAD1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000130816_DNMT1.pdf">ENSG00000130816_DNMT1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000131236_CAP1.pdf">ENSG00000131236_CAP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000131828_PDHA1.pdf">ENSG00000131828_PDHA1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000132406_TMEM128.pdf">ENSG00000132406_TMEM128</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000132475_H3F3B.pdf">ENSG00000132475_H3F3B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000132823_OSER1.pdf">ENSG00000132823_OSER1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000132824_SERINC3.pdf">ENSG00000132824_SERINC3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000133030_MPRIP.pdf">ENSG00000133030_MPRIP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000133561_GIMAP6.pdf">ENSG00000133561_GIMAP6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000133574_GIMAP4.pdf">ENSG00000133574_GIMAP4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000133639_BTG1.pdf">ENSG00000133639_BTG1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000133835_HSD17B4.pdf">ENSG00000133835_HSD17B4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000135046_ANXA1.pdf">ENSG00000135046_ANXA1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000135269_TES.pdf">ENSG00000135269_TES</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000135972_MRPS9.pdf">ENSG00000135972_MRPS9</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000136490_LIMD2.pdf">ENSG00000136490_LIMD2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000136628_EPRS.pdf">ENSG00000136628_EPRS</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000137078_SIT1.pdf">ENSG00000137078_SIT1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000137154_RPS6.pdf">ENSG00000137154_RPS6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000138172_CALHM2.pdf">ENSG00000138172_CALHM2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000138757_G3BP2.pdf">ENSG00000138757_G3BP2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000138794_CASP6.pdf">ENSG00000138794_CASP6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000139187_KLRG1.pdf">ENSG00000139187_KLRG1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000139193_CD27.pdf">ENSG00000139193_CD27</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000139644_TMBIM6.pdf">ENSG00000139644_TMBIM6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000140988_RPS2.pdf">ENSG00000140988_RPS2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000142227_EMP3.pdf">ENSG00000142227_EMP3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000143106_PSMA5.pdf">ENSG00000143106_PSMA5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000143333_RGS16.pdf">ENSG00000143333_RGS16</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000143575_HAX1.pdf">ENSG00000143575_HAX1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000143799_PARP1.pdf">ENSG00000143799_PARP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000144381_HSPD1.pdf">ENSG00000144381_HSPD1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000145050_MANF.pdf">ENSG00000145050_MANF</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000145779_TNFAIP8.pdf">ENSG00000145779_TNFAIP8</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000146112_PPP1R18.pdf">ENSG00000146112_PPP1R18</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000146386_ABRACL.pdf">ENSG00000146386_ABRACL</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000147403_RPL10.pdf">ENSG00000147403_RPL10</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000147604_RPL7.pdf">ENSG00000147604_RPL7</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000148303_RPL7A.pdf">ENSG00000148303_RPL7A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000149273_RPS3.pdf">ENSG00000149273_RPS3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000150753_CCT5.pdf">ENSG00000150753_CCT5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000152518_ZFP36L2.pdf">ENSG00000152518_ZFP36L2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000154473_BUB3.pdf">ENSG00000154473_BUB3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000155660_PDIA4.pdf">ENSG00000155660_PDIA4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000155959_VBP1.pdf">ENSG00000155959_VBP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000156508_EEF1A1.pdf">ENSG00000156508_EEF1A1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000157303_SUSD3.pdf">ENSG00000157303_SUSD3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000157916_RER1.pdf">ENSG00000157916_RER1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000161016_RPL8.pdf">ENSG00000161016_RPL8</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000161800_RACGAP1.pdf">ENSG00000161800_RACGAP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000162704_ARPC5.pdf">ENSG00000162704_ARPC5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000162777_DENND2D.pdf">ENSG00000162777_DENND2D</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000162894_FCMR.pdf">ENSG00000162894_FCMR</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163154_TNFAIP8L2.pdf">ENSG00000163154_TNFAIP8L2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163191_S100A11.pdf">ENSG00000163191_S100A11</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163219_ARHGAP25.pdf">ENSG00000163219_ARHGAP25</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163519_TRAT1.pdf">ENSG00000163519_TRAT1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163545_NUAK2.pdf">ENSG00000163545_NUAK2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163599_CTLA4.pdf">ENSG00000163599_CTLA4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000163682_RPL9.pdf">ENSG00000163682_RPL9</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000164104_HMGB2.pdf">ENSG00000164104_HMGB2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000165672_PRDX3.pdf">ENSG00000165672_PRDX3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000165929_TC2N.pdf">ENSG00000165929_TC2N</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000167257_RNF214.pdf">ENSG00000167257_RNF214</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000167658_EEF2.pdf">ENSG00000167658_EEF2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000167851_CD300A.pdf">ENSG00000167851_CD300A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000167996_FTH1.pdf">ENSG00000167996_FTH1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000168028_RPSA.pdf">ENSG00000168028_RPSA</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000168061_SAC3D1.pdf">ENSG00000168061_SAC3D1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000168209_DDIT4.pdf">ENSG00000168209_DDIT4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000168685_IL7R.pdf">ENSG00000168685_IL7R</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000169220_RGS14.pdf">ENSG00000169220_RGS14</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000169223_LMAN2.pdf">ENSG00000169223_LMAN2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000169756_LIMS1.pdf">ENSG00000169756_LIMS1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000170633_RNF34.pdf">ENSG00000170633_RNF34</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000170915_PAQR8.pdf">ENSG00000170915_PAQR8</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000171863_RPS7.pdf">ENSG00000171863_RPS7</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000172586_CHCHD1.pdf">ENSG00000172586_CHCHD1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000172977_KAT5.pdf">ENSG00000172977_KAT5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000173113_TRMT112.pdf">ENSG00000173113_TRMT112</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000173812_EIF1.pdf">ENSG00000173812_EIF1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000174444_RPL4.pdf">ENSG00000174444_RPL4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000174500_GCSAM.pdf">ENSG00000174500_GCSAM</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000174748_RPL15.pdf">ENSG00000174748_RPL15</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000175166_PSMD2.pdf">ENSG00000175166_PSMD2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000175334_BANF1.pdf">ENSG00000175334_BANF1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000175463_TBC1D10C.pdf">ENSG00000175463_TBC1D10C</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000177721_ANXA2R.pdf">ENSG00000177721_ANXA2R</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000177854_TMEM187.pdf">ENSG00000177854_TMEM187</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000178035_IMPDH2.pdf">ENSG00000178035_IMPDH2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000178537_SLC25A20.pdf">ENSG00000178537_SLC25A20</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000178896_EXOSC4.pdf">ENSG00000178896_EXOSC4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000178977_LINC00324.pdf">ENSG00000178977_LINC00324</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000179144_GIMAP7.pdf">ENSG00000179144_GIMAP7</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000179262_RAD23A.pdf">ENSG00000179262_RAD23A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000179933_C14orf119.pdf">ENSG00000179933_C14orf119</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000179950_PUF60.pdf">ENSG00000179950_PUF60</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000181163_NPM1.pdf">ENSG00000181163_NPM1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000181847_TIGIT.pdf">ENSG00000181847_TIGIT</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000182568_SATB1.pdf">ENSG00000182568_SATB1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000182718_ANXA2.pdf">ENSG00000182718_ANXA2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000182866_LCK.pdf">ENSG00000182866_LCK</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000183813_CCR4.pdf">ENSG00000183813_CCR4</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000184009_ACTG1.pdf">ENSG00000184009_ACTG1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000184432_COPB2.pdf">ENSG00000184432_COPB2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000184557_SOCS3.pdf">ENSG00000184557_SOCS3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000184588_PDE4B.pdf">ENSG00000184588_PDE4B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000184939_ZFP90.pdf">ENSG00000184939_ZFP90</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000185338_SOCS1.pdf">ENSG00000185338_SOCS1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000187257_RSBN1L.pdf">ENSG00000187257_RSBN1L</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000187514_PTMA.pdf">ENSG00000187514_PTMA</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000188846_RPL14.pdf">ENSG00000188846_RPL14</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000189067_LITAF.pdf">ENSG00000189067_LITAF</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000196230_TUBB.pdf">ENSG00000196230_TUBB</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000196329_GIMAP5.pdf">ENSG00000196329_GIMAP5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000196531_NACA.pdf">ENSG00000196531_NACA</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000197329_PELI1.pdf">ENSG00000197329_PELI1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000197471_SPN.pdf">ENSG00000197471_SPN</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000197958_RPL12.pdf">ENSG00000197958_RPL12</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198034_RPS4X.pdf">ENSG00000198034_RPS4X</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198130_HIBCH.pdf">ENSG00000198130_HIBCH</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198242_RPL23A.pdf">ENSG00000198242_RPL23A</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198763_MT-ND2.pdf">ENSG00000198763_MT-ND2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198805_PNP.pdf">ENSG00000198805_PNP</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198856_OSTC.pdf">ENSG00000198856_OSTC</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198938_MT-CO3.pdf">ENSG00000198938_MT-CO3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000204261_PSMB8-AS1.pdf">ENSG00000204261_PSMB8-AS1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000204386_NEU1.pdf">ENSG00000204386_NEU1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000204438_GPANK1.pdf">ENSG00000204438_GPANK1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000204619_PPP1R11.pdf">ENSG00000204619_PPP1R11</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000204628_RACK1.pdf">ENSG00000204628_RACK1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000205413_SAMD9.pdf">ENSG00000205413_SAMD9</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211450_SELENOH.pdf">ENSG00000211450_SELENOH</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211750_TRBV24-1.pdf">ENSG00000211750_TRBV24-1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211786_TRAV8-2.pdf">ENSG00000211786_TRAV8-2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211788_TRAV13-1.pdf">ENSG00000211788_TRAV13-1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211789_TRAV12-2.pdf">ENSG00000211789_TRAV12-2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211793_TRAV9-2.pdf">ENSG00000211793_TRAV9-2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211801_TRAV21.pdf">ENSG00000211801_TRAV21</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211804_TRDV1.pdf">ENSG00000211804_TRDV1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211806_TRAV25.pdf">ENSG00000211806_TRAV25</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000211810_TRAV29DV5.pdf">ENSG00000211810_TRAV29DV5</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000213203_GIMAP1.pdf">ENSG00000213203_GIMAP1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000213626_LBH.pdf">ENSG00000213626_LBH</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000234745_HLA-B.pdf">ENSG00000234745_HLA-B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000235532_LINC00402.pdf">ENSG00000235532_LINC00402</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000240065_PSMB9.pdf">ENSG00000240065_PSMB9</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000241657_TRBV11-2.pdf">ENSG00000241657_TRBV11-2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000242485_MRPL20.pdf">ENSG00000242485_MRPL20</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000243749_TMEM35B.pdf">ENSG00000243749_TMEM35B</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000243927_MRPS6.pdf">ENSG00000243927_MRPS6</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000257599_OVCH1-AS1.pdf">ENSG00000257599_OVCH1-AS1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000277734_TRAC.pdf">ENSG00000277734_TRAC</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000278030_TRBV7-9.pdf">ENSG00000278030_TRBV7-9</a> <a href="Fig/STEP3//Fig_DEG_example_TIGIT_TIGIT.pdf">TIGIT_TIGIT</a></p>
</div>
<div id="additional-genesproteins" class="section level2">
<h2>4. Additional genes/proteins</h2>
<pre class="r"><code>.add &lt;- c(&quot;CD279&quot;,&quot;PDCD1&quot;,&quot;LAG3&quot;,&quot;CD223&quot;,&quot;HAVCR2&quot;,&quot;TIM3&quot;,&quot;CD366&quot;,&quot;TOX&quot;,&quot;TOX2&quot;)

additional.genes &lt;-
    hc.ms.deg[hgnc_symbol %in% .add, .(gene)] %&gt;%
    unique() %&gt;%
    unlist(use.names=F)</code></pre>
<p><a href="Fig/STEP3//Fig_DEG_example_CD279_CD279.pdf">CD279_CD279</a> <a href="Fig/STEP3//Fig_DEG_example_CD366_CD366.pdf">CD366_CD366</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000089692_LAG3.pdf">ENSG00000089692_LAG3</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000124191_TOX2.pdf">ENSG00000124191_TOX2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000135077_HAVCR2.pdf">ENSG00000135077_HAVCR2</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000188389_PDCD1.pdf">ENSG00000188389_PDCD1</a> <a href="Fig/STEP3//Fig_DEG_example_ENSG00000198846_TOX.pdf">ENSG00000198846_TOX</a></p>
</div>
<div id="appendix-prdm1-short-vs.-long" class="section level2">
<h2>5. Appendix: PRDM1 short vs. long</h2>
<pre class="r"><code>.total.cols &lt;- fread(.deg.data$col, col.names = &quot;tag&quot;, header=F)

## 1. Write down PRDM1 short and long as .mtx
.hdr &lt;- &quot;result/step3/deg/prdm1_hc_ms&quot;
if.needed(fileset.list(.hdr), {

    .dt &lt;- fread(&quot;data/PRDM1/PRDM1_SL.csv.gz&quot;)
    .dt[, tag := barcode %&amp;% &quot;-1&quot; %&amp;% &quot;_&quot; %&amp;% `batch`]

    .out &lt;- .dt[(PRDM1_short + PRDM1_long &gt; 0) &amp;
                `tag` %in% .total.cols$tag,
                .(tag, PRDM1_short, PRDM1_long)]

    .mat &lt;- .out[, .(PRDM1_short, PRDM1_long)] %&gt;%
        as.matrix() %&gt;%
        t() %&gt;% 
        Matrix::Matrix(sparse=TRUE)

    mmutilR::write.sparse(.mat,
                          c(&quot;PRDM1_short&quot;, &quot;PRDM1_long&quot;),
                          .out$tag,
                          .hdr)
})


prdm1.data &lt;- fileset.list(.hdr)

.file &lt;- &quot;result/step3/deg/prdm1_hc_ms.rds&quot;

if.needed(.file, {

    .pca &lt;- rcpp_mmutil_svd(mtx_file = .deg.data$mtx,
                            RANK = 15,
                            TAKE_LN = T)

    .prdm1.cols &lt;- fread(prdm1.data$col, col.names = &quot;tag&quot;, header=F)

    V &lt;- .pca$V[match(.prdm1.cols$tag, .total.cols$tag),
              , drop = F]

    .prdm1.cols[, c(&quot;barcode&quot;, &quot;batch&quot;) := tstrsplit(`tag`, split=&quot;_&quot;)]
    .prdm1.cols[, barcode := gsub(pattern=&quot;-1&quot;, replacement = &quot;&quot;, `barcode`)]
    .prdm1.cols[, batch := as.integer(batch)]

    .matched &lt;- .prdm1.cols %&gt;%
        left_join(annot.dt) %&gt;% 
        left_join(.hash.info)

    .stat &lt;- rcpp_mmutil_aggregate(mtx_file = prdm1.data$mtx, 
                                   row_file = prdm1.data$row,
                                   col_file = prdm1.data$col, 
                                   r_cols = .matched$tag,
                                   r_indv = .matched$subject,
                                   r_annot = .matched$celltype,
                                   r_lab_name = c(&quot;mTreg&quot;,&quot;mTconv&quot;,&quot;nTreg&quot;,&quot;nTconv&quot;),
                                   r_trt = .matched$disease,
                                   r_V = V, knn = 50, IMPUTE_BY_KNN = T,
                                   NUM_THREADS = 16)

    saveRDS(.stat, .file)
})

.deg.stat &lt;- readRDS(.file)

.cts &lt;- unique(annot.dt$celltype)
.indvs &lt;- unique(annot.dt$subject)

.hc.ms.dt &lt;-
    list(tot = sort.col(.deg.stat$sum, .cts, .indvs),
         cfa = sort.col(.deg.stat$resid.ln.mu, .cts, .indvs),
         cfa.sd = sort.col(.deg.stat$resid.ln.mu.sd, .cts, .indvs)) %&gt;%
    combine.statistics()

.hc.ms.dt[, c(&quot;subject&quot;,&quot;celltype&quot;) := tstrsplit(`Var2`,split=&quot;_&quot;)]
.hc.ms.dt[, disease := substr(`subject`,1,2)]
.hc.ms.dt[, gene := `Var1`]
.hc.ms.dt[, `Var1` := NULL]
.hc.ms.dt[, `Var2` := NULL]

hc.ms.deg &lt;- summarize.deg(.hc.ms.dt)</code></pre>
<p><img src="Fig/STEP3/Fig_SL_DEG_PRDM1_long-1.png" /><!-- --></p>
<p><a href="Fig/STEP3//Fig_SL_DEG_PRDM1_long.pdf">PDF</a></p>
<p><img src="Fig/STEP3/Fig_SL_DEG_PRDM1_short-1.png" /><!-- --></p>
<p><a href="Fig/STEP3//Fig_SL_DEG_PRDM1_short.pdf">PDF</a></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
