---
title: "Step 3: differential Expression analysis"
date: "`r Sys.Date()`"
author: Yongjin Park
bibliography: "MS_Ref.bib"
output:
  html_document:
    toc: true
    keep_md: true
    self_contained: true
---

```{r include = F}
## unlink("result/step3",recursive = T)
## unlink("Fig/STEP3",recursive = T)
## unlink("Tab/DEG_MS_vs_HC.txt.gz")
```

```{r GlobOpt, include=FALSE}
fig.dir = 'Fig/STEP3/'
knitr::opts_knit$set(eval.after = 'fig.cap')
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.path = fig.dir)
knitr::opts_chunk$set(fig.width=8, fig.height=8)
options(stringsAsFactors = FALSE)
FIG.CAP = '**Fig.**'
library(tidyverse)
library(data.table)
library(patchwork)
library(mmutilR)
source("Util-rmd.R")
source("Util-alg.R")
.mkdir(fig.dir)
```

```{r read_data_info}
.data <- fileset.list("result/step1/matrix")
.mkdir("result/step3")
```

```{r read_hash_data}
.features <- readLines(.data$row)
.hash <- .features[str_detect(.features, "Hash")]
.hash.hdr <- "result/step1/hash"
.hash.data <- fileset.list(.hash.hdr)
.hash.info <- read.hash(.hash.data)
```

## 1. MS vs. HC at the major cell type level

```{r read_annotation}
final.cell.type <-
     fread("Tab/step2_cell_type.txt.gz") %>%
     left_join(.hash.info) %>%
     na.omit()
```

```{r include = F}
sort.col <- function(.mat, .cts, .indvs){

    .fun <- function(ct) paste0(.indvs, "_", ct)

    .samples <-
        lapply(.cts, .fun) %>%
        do.call(what=c) %>%
        sort()

    .idx <- match(.samples, colnames(.mat))
    .idx <- .idx[!is.na(.idx)]
    .mat[, .idx, drop = FALSE]
}

combine.statistics <- function(.result){

    ret <- lapply(names(.result), function(v) {
        .mat <- .result[[v]]
        if(!is.matrix(.mat)) return(NULL)
        as.data.table(reshape2::melt(.mat, value.name = v))
    })

    out <- ret[[1]]
    for(j in 2:length(ret)){
        if(is.null(out)) {
            out <- ret[[j]]
        } else if(!is.null(ret[[j]])) {
            out <- merge(out, ret[[j]])
        }
    }
    return(out)
}

summarize.deg <- function(.dt, tot.cutoff = 10, disease.name = "MS", control.name = "HA") {

    ## adjust potential bias
    .dt[, b := 0]
    .dt[`tot` >= tot.cutoff, b := mean(cfa), by = .(gene, celltype)]
    .dt[, cfa := cfa - b]

    .sd <- .dt[`tot` >= tot.cutoff,
               .(min.sd = sd(cfa) / sqrt(.N)),
               by = .(gene, celltype)]

    .dt <- .dt %>%
        left_join(.sd) %>%
        filter(`tot` >= tot.cutoff) %>%
        mutate(w = 1/cfa.sd^2) %>%
        as.data.table

    ## average disease effect on the disease subjects
    ADD.dt <-
        .dt[disease == disease.name & `tot` >= tot.cutoff,
               .(ADD = sum(cfa * w) / sum(w),
                 ADD.se = sqrt(1/sum(w))),
               by = .(gene, celltype)]

    ## average disease effect on the control subjects
    ADC.dt <-
        .dt[disease == control.name & `tot` >= tot.cutoff,
               .(ADC = -sum(cfa * w)/ sum(w),
                 ADC.se = sqrt(1/sum(w))),
               by = .(gene, celltype)]

    ## average disease effect
    ADE.dt <-
        .dt[`tot` >= tot.cutoff] %>%
        mutate(cfa.s = if_else(disease == disease.name, 1, -1) * cfa) %>%
        mutate(tot.s = if_else(disease == disease.name, 1, -1) * tot) %>%
        as.data.table %>%
        (function(xx) {
            xx[,
               .(ADE = sum(cfa.s * w) / sum(w),
                 ADE.se = sqrt(1/sum(w)),
                 tot = sum(tot)),
               by = .(gene, celltype)
               ]
        }) %>%
        mutate(z = ADE/ADE.se) %>%
        mutate(pv = 2 * pnorm(abs(z), lower.tail=FALSE)) %>%
        group_by(celltype) %>%
        mutate(fwer = p.adjust(pv, "holm")) %>%
        mutate(fdr = p.adjust(pv, "fdr")) %>%
        ungroup() %>%
        as.data.table

    left_join(ADE.dt, ADD.dt) %>% left_join(ADC.dt) %>% as.data.table
}

format.deg.tab <- function(deg.tab, fig.hdr){

    .deg.cast <- function(.dt, .var, .fun = mean){
        .dt %>%
            mutate(k = .var %&% "_" %&% celltype) %>%
            dcast(ensembl_gene_id + hgnc_symbol ~ k,
                  value.var = .var,
                  fun.aggregate=.fun)
    }

    ret <-
        .deg.cast(deg.tab, "z", .fun = function(x) num.round(mean(x,na.rm=T))) %>%
        left_join(.deg.cast(deg.tab, "pv", .fun = function(x) num.sci(min(x,na.rm=T)))) %>%
        left_join(.deg.cast(deg.tab, "fwer", .fun = function(x) num.sci(min(x,na.rm=T)))) %>%
        as.data.table()

    ret[, .file := fig.hdr %&%
              ensembl_gene_id %&% "_" %&%
              hgnc_symbol %&% ".pdf"]

    ret[, .link := "[PDF](" %&% .file %&% ") "]
    ret[, .file := NULL]

    ret[!str_starts(ensembl_gene_id, "ENSG"), ensembl_gene_id := ""]

    return(ret)
}
```

### a. Prepare data to just include cells annotated for their cell types and phenotypes

```{r prepare_deg_data}
.mkdir("result/step3/deg/")
.deg.data <- fileset.list("result/step3/deg/hc_ms")
if.needed(.deg.data, {
    .deg.data <-
        rcpp_mmutil_copy_selected_columns(.data$mtx,
                                          .data$row,
                                          .data$col,
                                          unique(final.cell.type$tag),
                                          "result/step3/deg/hc_ms")
})
```
### b. Gene Q/C to remove genes with too many zeros

```{r}
nnz.cutoff <- 500
cv.cutoff <- 1.0
```

```{r gene_score_for_qc}
.scores <- rcpp_mmutil_compute_scores(.deg.data$mtx,
                                      .deg.data$row,
                                      .deg.data$col)

row.scores <- setDT(.scores$row) %>%
    rename(gene = name) %>%                      # 
    filter(!str_detect(`gene`,"[Hh]ashtag")) %>% # remove hashtag
    as.data.table() %>%
    parse.gene()
```

```{r Fig_gene_scores, fig.width=4, fig.height=4}
.genes.interest <- c("PRDM1", "SGK1", "LBH", "DDIT4","GIMAP7","ID3","ICAM2")
.lab.dt <- row.scores[hgnc_symbol %in% .genes.interest]

plt <-
    ggplot(row.scores, aes(log1p(nnz), log1p(cv))) +
    theme_classic() +
    stat_density_2d(geom = "raster",
                    aes(fill = after_stat(sqrt(density))),
                    show.legend=F,
                    contour=F) +
    scale_fill_viridis_c() +
    geom_vline(xintercept = log1p(nnz.cutoff), col=2, lty = 2) +
    geom_hline(yintercept = log1p(cv.cutoff), col=2, lty = 2) +
    stat_density2d(linewidth=.2, colour="white") +
        geom_point(data=.lab.dt, pch=19, col="white", stroke=0, size=2) +
        geom_text_repel(aes(label=hgnc_symbol), data=.lab.dt, size=4, min.segment.length = .01, segment.color="white", color="white") +
    scale_x_continuous("Number of Non zeros", labels = function(x) num.sci(round(10^x))) +
    scale_y_continuous("Coefficient of variation", labels = function(x) num.sci(10^x))

print(plt)
```

```{r results="asis", echo = F}
.file <- fig.dir %&% "/Fig_gene_scrores.pdf"
.gg.save(filename = .file, plot = plt, width=4, height=4)
```

```{r filter_deg_data_qc}
qc.features <- row.scores[nnz > nnz.cutoff & cv > cv.cutoff]

.qc.hdr <- "result/step3/deg/qc_mc_ms"
.qc.data <- fileset.list(.qc.hdr)

if.needed(.qc.data, {
    .qc.data <-
        rcpp_mmutil_copy_selected_rows(.deg.data$mtx,
                                       .deg.data$row,
                                       .deg.data$col,
                                       qc.features$gene,
                                       .qc.hdr)
})
```

```{r estimate_deg_statistics}
.file <- "result/step3/hc_ms.rds"
if.needed(.file,
{
    ## cell types
    .celltype <-
        final.cell.type %>%
        select(tag, celltype) %>%
        as.data.frame()
    ## cell -> individual
    .cell2indv <- final.cell.type %>%
        left_join(.hash.info) %>%
        select(tag, subject) %>%
        unique %>%
        as.data.frame()
    ## individual -> disease
    .indv2exp <- .cell2indv %>%
        select(subject) %>%
        mutate(disease = substr(`subject`, 1, 2)) %>%
        as.data.frame()

    .deg.stat <-
        make.cocoa(.qc.data, .celltype, .cell2indv, .indv2exp,
                   knn = 100, .rank = 15, .take.ln = TRUE,
                   impute.by.knn = TRUE, num.threads = 16)
    ## save for computed results
    saveRDS(.deg.stat, .file)
})
.deg.stat <- readRDS(.file)

.cts <- unique(final.cell.type$celltype)
.indvs <- unique(final.cell.type$subject)

.hc.ms.dt <-
    list(tot = sort.col(.deg.stat$sum, .cts, .indvs),
         cfa = sort.col(.deg.stat$resid.ln.mu, .cts, .indvs),
         cfa.sd = sort.col(.deg.stat$resid.ln.mu.sd, .cts, .indvs)) %>%
    combine.statistics() %>%
    na.omit() %>%
    as.data.table() %>%
    (function(x) {
        x[, c("sample", "celltype") := tstrsplit(as.character(Var2), split="_")];
        x[, disease := substr(`sample`, 1, 2)];
        x[, gene := as.character(Var1)];
        x
    }) %>%
    dplyr::select(-Var1, -Var2) %>%
    as.data.table()

hc.ms.deg <-
    summarize.deg(.hc.ms.dt, tot.cutoff = 5) %>%
    as.data.table() %>%
    parse.gene()
```

```{r Fig_DEG_hist, fig.width=5, fig.height=3}
plt <-
    .gg.plot(hc.ms.deg, aes(pv, fill=`celltype`)) +
    geom_histogram(bins=100, linewidth=.1, color="black") +
    xlab("p-value") +
    scale_fill_brewer(palette = "Paired")

print(plt)
```

```{r results="asis", echo = F}
.file <- fig.dir %&% "/Fig_DEG_hist.pdf"
.gg.save(filename = .file, plot = plt, width=5, height=3)
```

```{r ms_vs_hc_results, results="asis", echo = FALSE}
.mkdir("Tab")
.file <- "Tab/DEG_MS_vs_HC.txt.gz"
if.needed(.file, {
    fwrite(hc.ms.deg, .file, sep="\t")
})
cat("[**DOWNLOAD:** DEG MS vs HC]("%&% .file %&% ")\n\n")
```

### c. Comparison with the bulk DEG results

```{r read_bulk_deg_results}
read.bulk <- function(.file) {
    fread(.file) %>%
        rename(hgnc_symbol = gene_name) %>%
        rename(bulk.t = t, bulk.pv = P.Value, bulk.qv = adj.P.Val) %>%
        dplyr::select(hgnc_symbol, starts_with("bulk"))
}

.file <- "data/DEG/20180513/deg.ms.hc.treg.mem.exvivo.sex_covar.ruv.txt"
bulk.mTreg.dt <- read.bulk(.file)

.file <- "data/DEG/20180513/deg.ms.hc.teff.mem.exvivo.sex_covar.ruv.txt"
bulk.mTconv.dt <- read.bulk(.file)
```


```{r include = F}
plot.bulk.sc <- function(.sc, .bulk,
                         .key.genes = c("PRDM1","SGK1"),
                         fwer.cutoff = .05,
                         tot.cutoff = 500,
                         qv.cutoff = .2,
                         n.top = 7) {

    .sc.uniq <- .sc[order(pv), tail(.SD, 1), by = .(hgnc_symbol)]

    .dt <-
        left_join(.sc.uniq, .bulk) %>%
        filter(str_starts(`ensembl_gene_id`, "ENSG")) %>%
        na.omit() %>%
        mutate(sc.z = z) %>%
        na.omit %>%
        arrange(abs(sc.z)) %>%
        as.data.table

    .show <-
        .dt[(`fwer` < `fwer.cutoff` &
             `bulk.qv` < qv.cutoff &
             `tot` > tot.cutoff)] %>%
        group_by(sign(sc.z)) %>%
        top_n(n.top, -log10(pv)) %>%
        select(hgnc_symbol) %>%
        unlist()

    .dt.show <- .dt[hgnc_symbol %in% c(.show, .key.genes)]

    .dt.cor <- .dt[abs(bulk.t) > 2]
    rr <- cor(.dt.cor$z, .dt.cor$bulk.t, method="spearman")
    r.pv <- cor.test(.dt$z, .dt$bulk.t, method="spearman")$p.value

    .dt.pos <- .dt.show[z > 0]
    .dt.neg <- .dt.show[z < 0]

    y.lim <- range(c(.dt.neg$sc.z, .dt.pos$sc.z))

    plt <-
        ggplot(.dt, aes(x = bulk.t, y = sc.z)) +
        theme_classic() +
        xlab("bulk RNA-seq DEG effect") +
        ylab("scRNA-seq DEG effect") +
        scale_y_continuous(limits = y.lim) +
        ggrastr::rasterise(
                     geom_point(aes(size = -log10(pmax(pv, bulk.pv))),
                                stroke=0, pch=19, colour="gray40",
                                size=1, alpha = .5),
                     dpi=300) +
        geom_vline(xintercept = 0, linewidth = .1) +
        geom_hline(yintercept = 0, linewidth = .1) +
        geom_smooth(aes(bulk.t, sc.z), method="lm", se=FALSE, colour="red", lty=1, formula = y ~ x, linewidth=.5)

    .aes <- aes(label = hgnc_symbol %&% "\n(" %&% num.sci(bulk.pv) %&% ")\n" %&% "(" %&% num.sci(fwer) %&% ")")

    if(nrow(.dt.pos) > 0) {
        plt <- plt +
            geom_point(data=.dt.pos, pch=21, fill="#ef8a62") +
            ggrepel::geom_text_repel(data=.dt.pos, .aes,
                                     size = 2, vjust=0, hjust=0,
                                     min.segment.length = .05,
                                     segment.size = .5,
                                     segment.color = "#ef8a62",
                                     max.overlaps = 100)
    }

    if(nrow(.dt.neg) > 0) {
        plt <- plt +
            geom_point(data=.dt.neg, pch=21, fill="#67a9cf") +
            ggrepel::geom_text_repel(data=.dt.neg, .aes,
                                     size = 2, vjust=1, hjust=1,
                                     min.segment.length = .05,
                                     segment.size = .5,
                                     segment.color = "#67a9cf",
                                     max.overlaps = 100)
    }

    plt <- plt +
        ggtitle("Cor=" %&%
                num.round(rr) %&%
                " (bulk |t| > 2), p=" %&% num.sci(r.pv) %&%
                ", N=" %&% num.int(nrow(.dt)))

    list(plt=plt, cor = rr, cor.pv = r.pv)
}
```

#### Memory Treg genes

```{r include = F}
.sc <- hc.ms.deg[celltype == "mTreg"]

.bulk <- bulk.mTreg.dt

.out <- plot.bulk.sc(.sc,
                     .bulk,
                     .key.genes = c("PRDM1","SGK1","LBH","CD6","JUNB"),
                     fwer.cutoff = .01,
                     qv.cutoff = .2,
                     n.top = 5)
```

##### Correlation `r num.round(.out$cor)` with p-value = `r num.sci(.out$cor.pv)`

```{r Fig_DEG_comparison_mTreg, fig.width=4, fig.height=7, echo = FALSE, results="asis"}
print(.out$plt)
.file <- fig.dir %&% "/Fig_DEG_comparison_mTreg.pdf"
.gg.save(filename = .file, plot = .out$plt, width=4, height=7)
```

* Here, we computed correlation between the bulk and scRNA-seq results after selecting on the DEGs marginally significant in the bulk analysis ($|t| > 2$).

* Two numbers below the marked genes: first, we show p-value of the bulk analysis; second, we show Bonferroni-adjusted p-values of the scRNA-seq analysis.

#### Memory Tconv genes

```{r}
.sc <- hc.ms.deg[celltype == "mTconv"]

.bulk <- bulk.mTconv.dt

.out <- plot.bulk.sc(.sc,
                     .bulk,
                     .key.genes = c("PRDM1","SGK1","LBH","CD6","JUNB"),
                     fwer.cutoff = .01,
                     qv.cutoff = .2,
                     n.top = 5)
```

#### Correlation `r num.round(.out$cor)` with p-value = `r num.sci(.out$cor.pv)`

```{r Fig_DEG_comparison_mTconv, fig.width=4, fig.height=7, echo = FALSE, results="asis"}
.file <- fig.dir %&% "/Fig_DEG_comparison_mTconv.pdf"
print(.out$plt)
.gg.save(filename = .file, plot = .out$plt, width=4, height=7)
```

* Here, we computed correlation between the bulk and scRNA-seq results after selecting on the DEGs marginally significant in the bulk analysis ($|t| > 2$).

* Two numbers below the marked genes: first, we show p-value of the bulk analysis; second, we show Bonferroni-adjusted p-values of the scRNA-seq analysis.

### Found `r num.int(nrow(unique(hc.ms.deg[fwer < .01, .(gene)])))` unique genes strongly perturbed by MS with FWER 5%

* Up-regulated: `r num.int(nrow(unique(hc.ms.deg[fwer < .05 & z > 0, .(gene)])))`

* Down-regulated:  `r num.int(nrow(unique(hc.ms.deg[fwer < .05 & z < 0, .(gene)])))`

* Total pairs of genes and cell types: `r num.int(nrow(hc.ms.deg))`

```{r}
count.deg <- function(.dt, fwer.cutoff = .01) {
    .dt[fwer < fwer.cutoff &
        sign(ADD) == sign(ADE) &
        sign(ADC) == sign(ADE),
        .(n = .N),
        by = .(celltype,
               direction = if_else(z > 0, "up", "down"))
        ] %>%
        mutate(direction = factor(direction, c("up", "down"))) %>%
        group_by(celltype) %>%
        arrange(desc(direction)) %>%
        mutate(nc = cumsum(n)) %>%
        ungroup
}
```

```{r Fig_DEG_count, fig.width=2, fig.height=2}
.df <- count.deg(hc.ms.deg)

plt <-
    .gg.plot(.df, aes(x = celltype, y = n, group = direction, fill = direction)) +
    theme(legend.position = c(1,1)) +
    theme(legend.justification = c(1,1)) +
    xlab("cell type") + ylab("# DEG") +
    geom_bar(stat="identity") +
    geom_text(aes(y=nc, label=n)) +
    scale_fill_manual(values = c("#ef8a62","#67a9cf"))

print(plt)
```

```{r results="asis", echo = F}
.file <- fig.dir %&% "/Fig_DEG_count.pdf"
.gg.save(filename = .file, plot = plt, width=2, height=2)
```

## 2. Show genes with FWER $< 5 \times 10^{-2}$

```{r include = F}
show.each.gene <- function(g, .data, .stat, .cutoff = 2) {

    .dt.g <- na.omit(.data[gene == g]) %>% filter(tot > .cutoff)
    .stat.dt <- na.omit(.stat[gene == g]) %>% rename(total = tot)

    .dir <- .stat.dt[which.max(abs(.stat.dt$ADE)), .(ADE)] %>%
        unlist

    if(.dir > 0) {

        .dt <- .dt.g %>%
            left_join(.stat.dt, by = c("celltype", "gene")) %>%
            group_by(disease, celltype) %>%
            arrange(cfa) %>%
            mutate(j = 1:n()) %>%
            ungroup() %>%
            mutate(x = disease %&% "_" %&% j)

    } else {

        .dt <- .dt.g %>%
            left_join(.stat.dt, by = c("celltype", "gene")) %>%
            group_by(disease, celltype) %>%
            arrange(desc(cfa)) %>%
            mutate(j = 1:n()) %>%
            ungroup() %>%
            mutate(x = disease %&% "_" %&% j)

    }

    .dt <- .dt %>%
        mutate(.pv = if_else(is.na(pv), "1", num.sci(pv))) %>%
        mutate(.facet = celltype %&% "\n(" %&% .pv %&% ")") %>%
        mutate(ADD.z = ADD/ADD.se, ADC.z = ADC/ADC.se) %>%
        as.data.table

    .stat.dt <- .dt[, .(celltype,
                        ADE, ADE.se,
                        ADD, ADD.se,
                        ADC, ADC.se,
                        pv, .facet)] %>%
        unique

    .aes <- aes(x = x,
                y = cfa,
                ymin = cfa - cfa.sd,
                ymax = cfa + cfa.sd,
                fill = disease)

    .scale.y <- scale_y_continuous(labels = function(x) num.round(exp(x)))

    p1 <-
        .gg.plot(.dt, .aes) +
        theme(legend.position = "top") +
        facet_wrap(~.facet, nrow = 1, scales="free") +
        .scale.y +
        theme(axis.text.x = element_text(size=4)) +
        theme(axis.title.x = element_blank()) +
        geom_hline(yintercept = 0, linewidth = 1, colour = "gray") +
        geom_hline(data = .stat.dt, aes(yintercept = ADE), linewidth = 1, colour = "#F8766D") +
        geom_linerange(linewidth = .2) +
        geom_point(aes(size = tot), pch = 21, stroke = .2) +
        scale_size_continuous("total(k)", range=c(0.1, 2.5), labels = function(x) num.int(x/1e3)) +
        scale_fill_manual(values = c("gray90","#F8766D")) +
        xlab("") + ylab("fold change") + ggtitle(g)

    .temp.1 <- .stat.dt %>%
        dplyr::select(celltype, pv, ADE, ADD, ADC) %>%
        mutate(ADC = - ADC) %>%                  # flip the effect
        melt(id.vars=c("celltype", "pv"), value.name = "mu")

    .temp.2 <- .stat.dt %>%
        dplyr::select(celltype, ADE.se, ADD.se, ADC.se) %>%
        melt(id.vars="celltype", value.name = "se") %>%
        mutate(variable = str_remove(variable, ".se"))

    .temp <- left_join(.temp.1, .temp.2, by = c("celltype", "variable"))

    .aes <- aes(x = variable,
                y = mu,
                ymin = mu - 2 * se,
                ymax = mu + 2 * se)

    p2 <-
        .gg.plot(.temp, .aes) +
        theme(axis.text.x = element_text(angle=90, hjust = 1, vjust= .5)) +
        facet_wrap(~celltype, nrow = 1, scales="free") +
        .scale.y + ylab("fold change") + xlab("") +
        geom_hline(yintercept = 0, lty = 2, linewidth = .5, colour = "black") +
        geom_linerange(linewidth = .2) +
        geom_point(aes(fill = variable, stroke = .2, size = -log10(pv)), pch = 21) +
        scale_fill_brewer(palette = "Greens", guide="none") +
        scale_size_continuous("p-value", range=c(.5, 2.5), breaks=seq(0, 10, 2), labels = function(x) 10^(-x))

    plt <- (p1 / p2) + patchwork::plot_layout(heights = c(4,2))
    return(plt)
}
```

```{r}
.bulk.genes <-
    rbind(bulk.mTreg.dt[bulk.qv < .2, .(hgnc_symbol)],
          bulk.mTconv.dt[bulk.qv < .2, .(hgnc_symbol)]) %>%
    unique() %>%
    unlist()

hc.ms.deg[, c("ensembl_gene_id", "hgnc_symbol") := tstrsplit(`gene`, split="_")]

.genes.show <-
    hc.ms.deg[fwer < 0.05 &
              hgnc_symbol %in% .bulk.genes &
              sign(ADD) == sign(ADE) &
              sign(ADC) == sign(ADE)] %>%
    select(gene) %>%
    unique %>%
    unlist %>%
    as.character
```

```{r include = F}
library(kableExtra)
```

```{r incldue = F}
hc.ms.deg[, c("ensembl_gene_id", "hgnc_symbol") := tstrsplit(`gene`, split="_")]

.genes.show <-
    hc.ms.deg[fwer < .05 &
              sign(ADD) == sign(ADE) &
              sign(ADC) == sign(ADE)] %>%
    select(gene) %>%
    unique() %>%
    unlist() %>%
    as.character()
```

```{r Fig_DEG_example, include=F}
.mkdir(fig.dir %&% "/example/")
for(g in unique(.genes.show)){
    .file <- fig.dir %&% "example/Fig_DEG_example_" %&% g %&% ".pdf"
   if.needed(.file, {
       plt <- show.each.gene(g, .hc.ms.dt, hc.ms.deg)
       ggsave(filename = .file, plot = plt, width=5, height=3, useDingbats=F, units="in")
   })
}
```

```{r Tab_DEG_FWER5, results="asis", echo = FALSE}
fig.hdr <- fig.dir %&% "/example/Fig_DEG_example_"
deg.tab <- hc.ms.deg[gene %in% .genes.show]

format.deg.tab(deg.tab, fig.hdr) %>%
    kbl() %>%
    kable_paper("striped", full_width=F, html_font = "Helvetica") %>%
    column_spec(1, bold = T)
```

## 3. Additional genes/proteins

```{r include = F}
.add <- c("CD279","PDCD1","LAG3","CD223","HAVCR2","TIM3","CD366","TOX","TOX2", "SGK1")

additional.genes <-
    hc.ms.deg[hgnc_symbol %in% .add, .(gene)] %>%
    unique() %>%
    unlist(use.names=F)
```

```{r Fig_DEG_example2, include = FALSE}
.mkdir(fig.dir %&% "/example/")
for(g in additional.genes){
    .file <- fig.dir %&% "/example/Fig_DEG_example_" %&% g %&% ".pdf"
    if.needed(.file, {
        plt <- show.each.gene(g, .hc.ms.dt, hc.ms.deg)
        ggsave(filename = .file, plot = plt, width=5, height=3, useDingbats=F, units="in")
    })
    cat("[" %&% g %&% "](" %&% .file %&% ") ")
}
```

```{r Tab_DEG_extra, results="asis", echo = FALSE}
fig.hdr <- fig.dir %&% "/example/Fig_DEG_example_"
deg.tab <- hc.ms.deg[gene %in% additional.genes]

format.deg.tab(deg.tab, fig.hdr) %>%
    kbl() %>%
    kable_paper("striped", full_width=F, html_font = "Helvetica") %>%
    column_spec(1, bold = T)
```
