---
title: "Cell type annotation and topic modelling"
date: "`r Sys.Date()`"
author: Yongjin Park
bibliography: "reference.bib"
---


```{r GlobOpt, include = FALSE}
################################################################
fig.dir = 'Fig/STEP2/'
dir.create(fig.dir, recursive = TRUE, showWarnings = FALSE)
knitr::opts_knit$set(eval.after = 'fig.cap')
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.path = fig.dir)
knitr::opts_chunk$set(fig.width=8, fig.height=8)
options(stringsAsFactors = FALSE)
FIG.CAP = '**Fig.**'
library(data.table)
library(dplyr)
library(patchwork)
library(mmutilR)
source("Util-rmd.R")
.fread <- function(...) fread(..., header=FALSE)
.fwrite <- function(...) fwrite(..., quote=FALSE, sep="\t", row.names = FALSE, col.names = FALSE)
```

```{r include=FALSE}
plt.scatter.ct.2 <- function(.ct.show, .assign.tab, .mtx,
                             m1x = "CD25",
                             m1y = "CD127",
                             m2x = "CD45RO",
                             m2y = "CD45RA") {

    .idx <- data.table(tag = colnames(.mtx)) %>%
        mutate(j = 1:n()) %>% 
        left_join(.assign.tab[as.character(celltype) %in% .ct.show]) %>%
        select(j, celltype, prob)

    .dt <- data.table(m1x = .mtx["anti_" %&% m1x, .idx$j],
                      m1y = .mtx["anti_" %&% m1y, .idx$j],
                      m2x = .mtx["anti_" %&% m2x, .idx$j],
                      m2y = .mtx["anti_" %&% m2y, .idx$j],
                      prob = .idx$prob,
                      celltype = factor(.idx$celltype, .ct.show))
    .dt <- .dt[celltype %in% .ct.show]
    .med.dt <-
        .dt[m1x > 0 &
            m2x > 0 &
            m1y > 0 &
            m2y > 0,
            .(m1x = median(m1x),
                m1y = median(m1y),
                m2x = median(m2x),
                m2y = median(m2y)),
            by = .(celltype)]

    .add.lab <- function(.plt) {

        .lab <- function(x) num.int(10^x)
        .brk <- seq(0,5)

        .plt +
            theme_classic() +
            theme(axis.text = element_text(size=6)) +
            scale_x_continuous(breaks=.brk, labels=.lab) +
            scale_y_continuous(breaks=.brk, labels=.lab) +
            theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
            facet_grid(.~celltype) +
            stat_density_2d(geom = "raster",
                            aes(fill = after_stat(density)),
                            show.legend=F,
                            contour=F) +
            scale_fill_viridis_c() +
            stat_density2d(linewidth=.2, colour="white") +
            geom_point(data=.med.dt, colour=2, pch=3, size=1, stroke=1) +
            geom_abline(slope=1, linewidth=.2, colour="white")
    }
    .aes <- aes(x=log10(m1x), y=log10(m1y))
    p1 <- (ggplot(.dt[m1x > 0 & m1y > 0], .aes) %>% .add.lab) +
        xlab(m1x) + ylab(m1y)
    .aes <- aes(x=log10(m2x), y=log10(m2y))
    p2 <- (ggplot(.dt[m2x > 0 & m2y > 0], .aes) %>% .add.lab) +
        xlab(m2x) + ylab(m2y)
    p1/p2
}
```

# Round 1: Top-level cell type annotation

## Can we distinguish cell types by surface proteins?

```{r echo = FALSE}
.read.marker.file <- function(.file){
    x <- .fread(.file, col.names=c("feat","ct"))
    .xx <- x[, .(feat=list(feat)), by = .(ct)];
    .ret <- .xx$feat; names(.ret) <- .xx$ct;
    .ret
}
```

```{r}
.hdr <- "result/step1/matrix_final"
.data <- fileset.list(.hdr)
```

```{r select_proteins_only, echo = FALSE}
.proteins <-
    .fread(.data$row, col.names="feat") %>%
    filter(str_starts(feat, "anti_")) %>%
    unlist()
  
.mkdir("result/step2")
.prot.hdr <- "result/step2/prot"
.prot.data <- fileset.list(.prot.hdr)

if.needed(.prot.data, {
    .prot.data <-
        rcpp_mmutil_copy_selected_rows(.data$mtx,
                                       .data$row,
                                       .data$col,
                                       .proteins,
                                       .prot.hdr)
})
```


```{r run_bbknn_protein_data}
.file <- "result/step2/prot_bbknn.rds"
if.needed(.file, {
    
    .batches <- .fread(.prot.data$col, col.names="tag")
    .batches[, c("barcode","batch") := tstrsplit(tag, "_")]

    .bbknn <- rcpp_mmutil_bbknn_pca(.prot.data$mtx,
                                    .batches$batch,
                                    knn = 10, RANK = 20,
                                    EM_ITER = 20,
                                    TAKE_LN = TRUE)
    saveRDS(.bbknn, .file)
})
.bbknn <- readRDS(.file)
```

* nTconv : CD3+, CD4+, CD8-, CD25-/CD127+, CD45RA+/CD45RO-
* mTconv : CD3+, CD4+, CD8-, CD25-/CD127+, CD45RA-/CD45RO+
* nTreg : CD3+, CD4+, CD8-, CD25+/CD127-, CD45RA+/CD45RO-
* mTreg : CD3+, CD4+, CD8-, CD25+/CD127-, CD45RA-/CD45RO+

Run annotation purely based on marker genes:

```{r run_annotation_bbknn_adjusted}
.file <- "result/step2/prot_annot.txt.gz"
if.needed(.file, {
    
    .pos.markers <- .read.marker.file("marker/surface_round1_positive.txt")
    .neg.markers <- .read.marker.file("marker/surface_round1_negative.txt")

    .annot.out <-
        rcpp_mmutil_annotate_columns(
            pos_labels = list(p1=.pos.markers),
            r_neg_labels = list(n1=.neg.markers),
            row_file = .prot.data$row,
            col_file = .prot.data$col,
            r_U = .bbknn$U,
            r_D = .bbknn$D,
            r_V = .bbknn$factors.adjusted,
            EM_TOL = 1e-6,
            EM_ITER = 500)
    annot.dt <- setDT(.annot.out$annotation)
    .col <- c("tag", "celltype", "prob", "ln.prob")
    names(annot.dt) <- .col
    annot.dt[, c("barcode","batch") := tstrsplit(tag, split="_")]
    fwrite(annot.dt, .file)
})
annot.dt <- fread(.file)
```

Two-dimensional density plot on the normalized CD marker expressions.

```{r Fig_round1_cdmarker_bbknn, fig.width=6, fig.height=5, results="asis"}
U <- .bbknn$U
D <- .bbknn$D
V <- .bbknn$factors.adjusted

prot.mtx <- pmax(exp(sweep(U, 2, D, `*`) %*% t(V)) - 1, 0)
prot.mtx <- apply(prot.mtx, 2, function(x) x/pmax(sum(x),1) * 1e4)

rownames(prot.mtx) <- readLines(.prot.data$row)
colnames(prot.mtx) <- readLines(.prot.data$col)

.ct <- c("mTreg","nTreg","mTconv","nTconv")
plt <- plt.scatter.ct.2(.ct, annot.dt, prot.mtx)
print(plt)
.file <- fig.dir %&% "/Fig_round1_cdmarker_bbknn.pdf"
.gg.save(filename = .file, plot = plt, width=6, height=5)
```

Two-dimensional density plot on the raw CD marker concentrations.

```{r Fig_round1_cdmarker, fig.width=6, fig.height=5, results="asis"}
prot.raw.mtx <- read.dense(.prot.data$mtx)
rownames(prot.raw.mtx) <- readLines(.prot.data$row)
colnames(prot.raw.mtx) <- readLines(.prot.data$col)

.ct <- c("mTreg","nTreg","mTconv","nTconv")
plt <- plt.scatter.ct.2(.ct, annot.dt, prot.raw.mtx)
print(plt)
.file <- fig.dir %&% "/Fig_round1_cdmarker.pdf"
.gg.save(filename = .file, plot = plt, width=6, height=5)
```

## Will the same classification results hold in transcriptomic data?

```{r select_genes_only, echo = FALSE}
.hdr <- "result/step1/matrix_final"
.data <- fileset.list(.hdr)

.genes <-
    .fread(.data$row, col.names="feat") %>%
    filter(!str_starts(feat, "anti_")) %>%
    unlist()
  
.mkdir("result/step2")
.gene.hdr <- "result/step2/gene"
.gene.data <- fileset.list(.gene.hdr)

if.needed(.gene.data, {
    .gene.data <-
        rcpp_mmutil_copy_selected_rows(.data$mtx,
                                       .data$row,
                                       .data$col,
                                       .genes,
                                       .gene.hdr)
})
```


```{r run_bbknn_genes_only}
.file <- "result/step2/gene_bbknn.rds"

if.needed(.file, {

    .batches <- .fread(.gene.data$col, col.names="tag")
    .batches[, c("barcode","batch") := tstrsplit(tag, "_")]

    .bbknn <- rcpp_mmutil_bbknn_pca(.gene.data$mtx,
                                    .batches$batch,
                                    knn = 10, RANK = 30,
                                    EM_ITER = 20,
                                    TAKE_LN = TRUE,
                                    NUM_THREADS = 8)
    saveRDS(.bbknn, .file)
})
.bbknn <- readRDS(.file)
```

```{r}
run.bbknn.umap <- function(.bbknn, .cells, ...){

    .df <- data.frame(from = .bbknn$knn$src.index,
                      to = .bbknn$knn$tgt.index,
                      weight = .bbknn$knn$weight)

    G <- igraph::graph_from_data_frame(.df, directed=FALSE)

    nrepeat <- 20
    C <- list(modularity = 0)
    for(r in 1:nrepeat){
        c.r <- igraph::cluster_louvain(G, resolution = .75)
        message(paste("modularity: ", c.r$modularity, "\n"))
        if(r == 1 || max(c.r$modularity) > max(C$modularity)){
            C <- c.r
        }
    }

    .clust <- data.table(tag = .cells[as.integer(C$names)],
                         membership = C$membership)

    A <- igraph::as_adj(G)

    umap.A <- uwot::optimize_graph_layout(A, verbose = TRUE, n_sgd_threads = "auto", ...)

    .dt <- data.table(tag = .cells[as.integer(rownames(A))],
                      umap1 = umap.A[,1],
                      umap2 = umap.A[,2]) %>% 
        left_join(.clust, by = "tag")
}
```

1. Build a batch-balancing kNN graph

2. Louvain graph-based clustering

3. Construct UMAP the kNN backbone graph

```{r}
.file <- "result/step2/gene_bbknn_louvain.txt.gz"
.cells <- readLines(.gene.data$col)
if.needed(.file, {
    .bbknn.umap <- run.bbknn.umap(.bbknn, .cells, spread=3)
    fwrite(.bbknn.umap, .file)
})
.bbknn.umap <- fread(.file) %>%
    left_join(annot.dt, by = "tag") %>% 
    mutate(membership = as.factor(membership))
```

### Clustering patterns in scRNA-seq agrees with the classification results based on surface proteins

```{r Fig_bbknn_gene_only_umap, fig.width=8, fig.height=3, results="asis"}
p1 <-
    .gg.plot(.bbknn.umap, aes(umap1, umap2, color=membership)) +
    geom_point(stroke=0, alpha=.8, size=.5)
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300)

p2 <-
    .gg.plot(.bbknn.umap, aes(umap1, umap2, color=celltype)) +
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300) +
    scale_color_brewer(palette = "Paired")

plt <- p1 | p2
print(plt)
.file <- fig.dir %&% "/Fig_bbknn_gene_only_umap.pdf"
.gg.save(filename = .file, plot = plt, width=8, height=3)
```


```{r Fig_bbknn_gene_only_cluster, fig.width=5, fig.height=2, results="asis"}
.ct.mem <- .bbknn.umap[, .(nc = .N), by = .(celltype, membership)]
.ct.mem[, ntot := sum(`nc`), by = .(membership)]
.ct.mem[, pr := `nc`/`ntot`]

.df <- 
    .ct.mem %>%
    mutate(row = celltype, col = membership,
           weight = logit(pmax(pmin(pr, 1-1e-2), 1e-2))) %>% 
    order.pair(ret.tab = TRUE)
plt <- 
    .gg.plot(.df, aes(col, row, fill=weight, size = `nc`)) +
    xlab("Louvain clustering") +
    ylab("cell type") +
    theme(legend.position = "bottom") +
    geom_point(pch=22) +
    scale_size_continuous("#cells", range=c(0,5)) +
    scale_fill_distiller("proportion",
                         palette="PuRd", direction = 1,
                         labels=function(x) round(sigmoid(x), 2))
print(plt)
.file <- fig.dir %&% "/Fig_bbknn_gene_only_cluster.pdf"
.gg.save(filename = .file, plot = plt, width=5, height=2)
```

## Joint clustering analysis

```{r run_bbknn_full}
.file <- "result/step2/full_bbknn.rds"

if.needed(.file, {

    .batches <- .fread(.data$col, col.names="tag")
    .batches[, c("barcode","batch") := tstrsplit(tag, "_")]

    .bbknn <- rcpp_mmutil_bbknn_pca(.data$mtx,
                                    .batches$batch,
                                    knn = 10, RANK = 30,
                                    EM_ITER = 20,
                                    TAKE_LN = TRUE,
                                    NUM_THREADS = 8)
    saveRDS(.bbknn, .file)
})
.bbknn <- readRDS(.file)
```

### 1. Run annotation on the joint BBKNN data

```{r run_annotation_bbknn_adjusted_full}
.file <- "result/step2/prot_annot_full.txt.gz"
if.needed(.file, {
    
    .pos.markers <- .read.marker.file("marker/surface_round1_positive.txt")
    .neg.markers <- .read.marker.file("marker/surface_round1_negative.txt")

    .annot.out <-
        rcpp_mmutil_annotate_columns(
            pos_labels = list(p1=.pos.markers),
            r_neg_labels = list(n1=.neg.markers),
            row_file = .data$row,
            col_file = .data$col,
            r_U = .bbknn$U,
            r_D = .bbknn$D,
            r_V = .bbknn$factors.adjusted,
            EM_TOL = 1e-6,
            EM_ITER = 500)
    annot.dt <- setDT(.annot.out$annotation)
    .col <- c("tag", "celltype", "prob", "ln.prob")
    names(annot.dt) <- .col
    annot.dt[, c("barcode","batch") := tstrsplit(tag, split="_")]
    fwrite(annot.dt, .file)
})
annot.dt <- fread(.file)
```

Two-dimensional density plot on the raw CD marker concentrations.

```{r Fig_round1_cdmarker_full, fig.width=6, fig.height=5, results="asis"}
prot.raw.mtx <- read.dense(.prot.data$mtx)
rownames(prot.raw.mtx) <- readLines(.prot.data$row)
colnames(prot.raw.mtx) <- readLines(.prot.data$col)

.ct <- c("mTreg","nTreg","mTconv","nTconv")
plt <- plt.scatter.ct.2(.ct, annot.dt, prot.raw.mtx)
print(plt)
.file <- fig.dir %&% "/Fig_round1_cdmarker_full.pdf"
.gg.save(filename = .file, plot = plt, width=6, height=5)
```

### 2. Graph-based clustering

```{r run_louvain_full}
.file <- "result/step2/full_bbknn_louvain.txt.gz"
.cells <- readLines(.data$col)
if.needed(.file, {
    .bbknn.umap <- run.bbknn.umap(.bbknn, .cells, spread=3)
    fwrite(.bbknn.umap, .file)
})
.bbknn.umap <- fread(.file) %>%
    left_join(annot.dt, by = "tag") %>% 
    mutate(membership = as.factor(membership))
```

```{r Fig_bbknn_full_umap, fig.width=8, fig.height=3, results="asis"}
p1 <-
    .gg.plot(.bbknn.umap, aes(umap1, umap2, color=membership)) +
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300)

p2 <-
    .gg.plot(.bbknn.umap, aes(umap1, umap2, color=celltype)) +
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300) +
    scale_color_brewer(palette = "Paired")

plt <- p1 | p2
print(plt)
.file <- fig.dir %&% "/Fig_bbknn_full_umap.pdf"
.gg.save(filename = .file, plot = plt, width=8, height=3)
```


```{r Fig_bbknn_full_cluster, fig.width=5, fig.height=2, results="asis"}
.ct.mem <- .bbknn.umap[, .(nc = .N), by = .(celltype, membership)]
.ct.mem[, ntot := sum(`nc`), by = .(membership)]
.ct.mem[, pr := `nc`/`ntot`]

.df <- 
    .ct.mem %>%
    mutate(row = celltype, col = membership,
           weight = logit(pmax(pmin(pr, 1-1e-2), 1e-2))) %>% 
    order.pair(ret.tab = TRUE)

plt <-
    .gg.plot(.df, aes(col, row, fill=weight, size = `nc`)) +
    xlab("Louvain clustering") +
    ylab("cell type") +
    theme(legend.position = "bottom") +
    geom_point(pch=22) +
    scale_size_continuous("#cells", range=c(0,5)) +
    scale_fill_distiller("proportion",
                         palette="PuRd", direction = 1,
                         labels=function(x) round(sigmoid(x), 2))
print(plt)
.file <- fig.dir %&% "/Fig_bbknn_full_cluster.pdf"
.gg.save(filename = .file, plot = plt, width=5, height=2)
```

### 3. Refine cell type assignments with the clustering results

```{r}
.valid.clust <- .ct.mem[order(pr, decreasing = T),
                        head(.SD, 1),
                        by = .(membership)]

.valid.celltype.cluster <-
    .valid.clust[pr > .5, .(membership, celltype)] %>%
    unique()

celltype.final <- 
    .bbknn.umap %>%
    rename(celltype.protein = celltype) %>%
    (function(x) left_join(.valid.celltype.cluster, x)) %>%
    na.omit()
```

```{r results="asis", echo=FALSE}
.file <- "Tab/step2_celltype_final.txt.gz"
.mkdir("Tab/")
if.needed(.file, fwrite(celltype.final, .file, sep = "\t", col.names = T))
cat("[**DOWNLOAD:** Cell type estimation]("%&% .file %&% ")\n\n")
```

```{r Fig_bbknn_final_umap, fig.width=8, fig.height=3, results="asis"}

p1 <-
    .gg.plot(celltype.final, aes(umap1, umap2, color=membership)) +
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300)

p2 <-
    .gg.plot(celltype.final, aes(umap1, umap2, color=celltype)) +
    ggrastr::rasterise(geom_point(stroke=0, alpha=.8, size=.5), dpi=300) +
    scale_color_brewer(palette = "Paired")

plt <- p1 | p2
print(plt)
.file <- fig.dir %&% "/Fig_bbknn_final_umap.pdf"
.gg.save(filename = .file, plot = plt, width=8, height=3)
```

### 4. Confirm by other marker genes/proteins

```{r}
.markers <-
    c("FOXP3", "ID3", "BACH2", "CXCR3", "PRDM1", "SGK1", "TCF7", "LEF1",
      "SELL", "IL2RA", "IL7R", "IKZF2", "CCR6", "CCR4", "CCR7", "CTLA4",
      "HLA-DRA", "anti_CD25", "anti_CD127", "anti_CD183", "anti_CD196",
      "anti_CD197", "anti_CD194", "anti_CD45RA", "anti_CD45RO",
      "anti_HLA") %>%
    unique
.marker.hdr <- "result/step2/marker/marker"
.mkdir(dirname(.marker.hdr))
.marker.data <- fileset.list(.marker.hdr)
if.needed(.marker.data, {
    .marker.data <-
        rcpp_mmutil_copy_selected_rows(.data$mtx,
                                       .data$row,
                                       .data$col,
                                       .markers,
                                       .marker.hdr)
})
```

```{r include = FALSE}
.tags <- readLines("result/step2/marker/marker.cols.gz")
    
.ct <- c("mTreg","nTreg","mTconv","nTconv")
.markers <- readLines(.marker.data$row)

marker.dt <-
    t(read.dense(.marker.data$mtx)) %>% 
    (function(x) { colnames(x) <- .markers; x }) %>%
    as.data.table() %>%
    mutate(tag = .tags) %>% 
    left_join(celltype.final[, .(tag, umap1, umap2, celltype)], by = "tag") %>%
    na.omit() %>% 
    as.data.table() %>%
    melt(id.vars = c("tag", "umap1", "umap2","celltype"),
         variable.name = "marker",
         value.name = "expr") %>%
    mutate(celltype = factor(celltype, .ct)) %>%
    as.data.table

plot.marker.umap <- function(.marker) {
    .dt <- marker.dt[marker == .marker]
    ## handle outliers for better visualization
    .cutoff <- quantile(.dt$expr, probs = .95, na.rm=TRUE)
    .cutoff <- max(.cutoff, 2.5)
    .aes <- aes(x = umap1,
                y = umap2,
                colour = expr)
    .dt <- .dt %>%
        mutate(expr = pmin(expr, .cutoff)) %>%
        arrange(expr)
    ret <-
        ggplot(.dt, .aes) +
        ggtitle(.marker) +
        xlab("UMAP 1") + ylab("UMAP 2") +
        theme_minimal() +
        ggrastr::rasterise(geom_point(stroke=.0, size=.7, alpha=.8), dpi=300) +
        theme(legend.key.width = unit(.2, "lines"),
              legend.key.height = unit(.5, "lines"),
              legend.title = element_text(size=6),
              legend.text = element_text(size=5))
    if(str_starts(.marker, "anti_")) {
        ret <- ret + scale_colour_distiller("", direction=1)
    } else {
        ret <- ret + scale_colour_distiller("", palette="RdPu", direction=1)
    }
    return(ret)
}
```

**Note: these plots show expression values without batch adjustment**

```{r Fig_marker, fig.width=3.2, fig.height=2.8, results="asis"}
.markers <- sort(as.character(unique(marker.dt$marker)))
for(g in .markers){
    plt <- plot.marker.umap(g)
    print(plt)
    .file <- fig.dir %&% "/Fig_marker_" %&% g %&% "_umap.pdf"
    .gg.save(filename = .file, plot = plt, width=3.2, height=2.8)
}
```

## Basic statistics for the first round annotation (`r num.int(nrow(celltype.final))` cells)

```{r include=FALSE}
.sum.stat.batch <- function(.dt) {
    .dt %>%
        group_by(batch, disease) %>%
        mutate(P = 100 * N/sum(N)) %>%
        arrange(desc(ct)) %>%
        mutate(Pc = cumsum(P)) %>%
        ungroup() %>%
        group_by(batch) %>%
        mutate(nb = sum(N)) %>%
        ungroup() %>%
        mutate(batch = "#" %&% batch %&% " (N=" %&% num.int(nb) %&% ")")
}
.sum.stat.tot <- function(.dt) {
    .dt %>%
        group_by(disease) %>%
        mutate(P = 100 * N/sum(N)) %>%
        arrange(desc(ct)) %>%
        mutate(Pc = cumsum(P)) %>%
        ungroup()
}
.plt.sum.stat <- function(.stat) {
    .gg.plot(.stat, aes(x=disease, y=P, group=ct, fill=celltype)) +
        facet_grid(.~batch) +
        geom_bar(stat="identity") +
        geom_text(aes(y=Pc, label=num.int(N)), size=2) +
        scale_fill_brewer(palette="Paired") +
        ylab("cell type composition (%)") +
        theme(legend.title=element_blank())
}
```


```{r}
.hash.info <- fread("data/cell_info_batch3-7.csv.gz") %>%
    rename(tag = cbid, subject = id) %>%
    select(-`V1`) %>%
    mutate(disease = substr(`subject`, 1, 2)) %>% 
    mutate(hash = gsub(pattern="[a-zA-Z\\-]+", replacement="", `hash`))

.sample.info <-
    readxl::read_xlsx("data/Hashing list MS Treg project.xlsx", 1) %>% 
    rename(Sample = `Cell type`) %>%
    mutate(hash = gsub("#","",`hash`))

.dt <-
    celltype.final %>%
    mutate(batch=as.integer(batch)) %>%
    left_join(.hash.info) %>%
    mutate(ct=factor(celltype, c("nTconv","mTconv","nTreg","mTreg"))) %>% 
    left_join(.sample.info)
```

## Combining all the experiments

```{r Fig_count_stat_tot, fig.height=3, fig.width=6, echo = FALSE, results="asis"}
.stat <-
    .dt[,
        .(N = .N),
        by=.(batch, celltype, ct, disease)] %>%
    .sum.stat.batch
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on all the samples")
print(plt)
.file <- fig.dir %&% "/Fig_count_stat_tot.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=6)
```

```{r Fig_count_merged_stat_tot, fig.height=3, fig.width=2.2, echo = FALSE, results="asis"}
.stat <-
    .dt[,
        .(N = .N),
        by=.(celltype, ct, disease)] %>%
    .sum.stat.tot %>%
    mutate(batch = "(N=" %&% num.int(sum(.stat$N)) %&% ")")
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on all the samples")
print(plt)
.file <- fig.dir %&% "/Fig_count_merged_stat_tot.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=2.2)
```

## Total CD4 samples

```{r Fig_count_stat_cd4, fig.height=3, fig.width=6, echo = FALSE, results="asis"}
.stat <-
    .dt[Sample == "Total CD4T cells",
        .(N = .N),
        by=.(batch, celltype, ct, disease)] %>%
    .sum.stat.batch
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on Total CD4 T-cells")
print(plt)
.file <- fig.dir %&% "/Fig_count_stat_cd4.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=6)
```


```{r Fig_count_merged_stat_cd4, fig.height=3, fig.width=2.2, echo = FALSE, results="asis"}
.stat <-
    .dt[Sample == "Total CD4T cells", .(N = .N), by=.(celltype, ct, disease)] %>%
    .sum.stat.tot %>%
    mutate(batch = "(N=" %&% num.int(sum(.stat$N)) %&% ")")
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on Total CD4 T-cells")
print(plt)
.file <- fig.dir %&% "/Fig_count_merged_stat_cd4.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=2.2)
```

## CD25 enriched samples

```{r Fig_count_stat_cd25enriched, fig.height=3, fig.width=6, echo = FALSE, results="asis"}
.stat <-
    .dt[Sample == "CD25+ CD4T cells",
        .(N = .N),
        by=.(batch, celltype, ct, disease)] %>%
    .sum.stat.batch
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on the CD25+ enriched")
print(plt)
.file <- fig.dir %&% "/Fig_count_stat_cd25enriched.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=6)
```

```{r Fig_count_merged_stat_cd25enriched, fig.height=3, fig.width=2.2, echo = FALSE, results="asis"}
.stat <-
    .dt[Sample == "CD25+ CD4T cells", .(N = .N), by=.(celltype, ct, disease)] %>%
    .sum.stat.tot %>%
    mutate(batch = "(N=" %&% num.int(sum(.stat$N)) %&% ")")
plt <- .plt.sum.stat(.stat) +
    ggtitle("Profiled on the CD25+ enriched")
print(plt)
.file <- fig.dir %&% "/Fig_count_merged_stat_cd25enriched.pdf"
.gg.save(filename = .file, plot = plt, height=3, width=2.2)
```
