---
title: "Differential Expression analysis for mTreg subtypes"
author: Yongjin Park
theme: jekyll-theme-minimal
date: "`r Sys.Date()`"
bibliography: "reference.bib"
---

```{r GlobOpt, include=FALSE}
fig.dir = 'Fig/STEP5/'
knitr::opts_knit$set(eval.after = 'fig.cap')
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.path = fig.dir)
knitr::opts_chunk$set(fig.width=8, fig.height=8)
options(stringsAsFactors = FALSE)
FIG.CAP = '**Fig.**'

library(tidyverse)
library(data.table)
library(patchwork)
source("Util-rmd.R")
.mkdir(fig.dir)
```

```{r helper_functions}
.fread <- function(...) fread(..., header=FALSE)
.fread.melt <-function(.file, .key, .val) {
    .ret = .fread(.file)
    .gather = colnames(.ret)
    .ret %>%
        mutate(g = 1:n()) %>%
        gather_(.key, .val, .gather) %>%
        as.data.table
}
.read.mean.sd <- function(.mean.file, .sd.file, .name) {
    .mean = .fread.melt(.mean.file, "col", .name)
    .sd = .fread.melt(.sd.file, "col", .name %&% ".sd")
    merge(.mean, .sd, by = c("g", "col"))
}
.fwrite <- function(...) fwrite(..., quote=FALSE, sep="\t", row.names = FALSE, col.names = FALSE)

.fwrite.safe <- function(.dt, .file, ...) {
    if(!file.exists(.file)) fwrite(.dt, .file, ...)
}

.read.mat <- function(...) .fread(...) %>% as.matrix
.read.vec <- function(...) .fread(...) %>% unlist(use.names=FALSE)


#' @param data.file E.g., "result/deg/strata.resid_mu.gz"
#' @param .name E.g., "resid.mu"
#' @param col.file E.g., "result/deg/strata.mu_cols.gz"
#' @param row.file E.g., "result/rna_final.rows.gz"
.read.stat <- function(data.file,
                       .name,
                       col.file,
                       row.file = "result/step5/rna_final.rows.gz") {

    .rows <- .fread(row.file)
    .cols <- .fread(col.file) %>% unlist

    .fread(data.file, col.names=.cols) %>%
        mutate(gene = unlist(.rows)) %>%
        melt(value.name = .name, id.vars = "gene", variable.name = "col") %>%
        as.data.table
}

read.deg.result <- function(hdr) {

    .fun <- function(.ext, .name) {
        .read.stat(hdr %&% .ext,
                   .name,
                   hdr %&% ".mu_cols.gz",
                   hdr %&% ".rows.gz")
    }

    .cfa <- .fun(".ln_resid_mu.gz", "cfa")
    .cfa.sd <- .fun(".ln_resid_mu_sd.gz", "cfa.sd")

    .mu <- .fun(".ln_obs_mu.gz", "mu")
    .mu.sd <- .fun(".ln_obs_mu_sd.gz", "mu.sd")

    .ret <-
        left_join(.cfa, .cfa.sd) %>%
        left_join(.mu) %>%
        left_join(.mu.sd) %>%
        as.data.table

    .ret[, c("batch","disease","celltype") := tstrsplit(col, split="_")]
    .ret[, batch := as.integer(batch)]
    .ret[, col := NULL]
    
    return(.ret)
}

summarize.deg <- function(.dt, expr.cutoff = 1e-4, disease.name = "MS", control.name = "HC") {

    ln.cutoff <- log(expr.cutoff)

    ## adjust potential bias
    .dt[mu >= ln.cutoff, b := mean(cfa), by = .(gene, celltype)]
    .dt[, cfa := cfa - b]

    .sd <- .dt[mu >= ln.cutoff,
               .(min.sd = sd(cfa) / sqrt(.N)),
               by = .(gene, celltype)]

    .dt <- .dt %>% 
        filter(mu >= ln.cutoff) %>% 
        left_join(.sd) %>%
        mutate(w = 1/(cfa.sd^2 + min.sd^2)) %>%
        as.data.table

    ## average disease effect on the disease subjects
    ADD.dt <-
        .dt[disease == disease.name & mu > ln.cutoff,
               .(ADD = sum(cfa * w) / sum(w),
                 ADD.se = sqrt(1/sum(w))),
               by = .(gene, celltype)]

    ## average disease effect on the control subjects
    ADC.dt <-
        .dt[disease == control.name & mu > ln.cutoff,
               .(ADC = -sum(cfa * w)/ sum(w),
                 ADC.se = sqrt(1/sum(w))),
               by = .(gene, celltype)]

    ## average disease effect
    ADE.dt <-
        .dt[mu > ln.cutoff] %>%
        mutate(cfa.s = if_else(disease == disease.name, 1, -1) * cfa) %>%
        mutate(mu.s = if_else(disease == disease.name, 1, -1) * mu) %>%
        as.data.table %>%
        (function(xx) {
            xx[,
               .(ADE = sum(cfa.s * w) / sum(w),
                 ADE.se = sqrt(1/sum(w))),
               by = .(gene, celltype)
               ]
        }) %>% 
        mutate(z = ADE/ADE.se) %>% 
        mutate(pv = 2 * pnorm(abs(z), lower.tail=FALSE)) %>%
        mutate(fwer = p.adjust(pv, "holm")) %>%
        as.data.table

    left_join(ADE.dt, ADD.dt) %>% left_join(ADC.dt) %>% as.data.table
}

plot.bulk.sc <- function(.sc, .bulk, fwer.cutoff = .5, qv.cutoff = .5) {

    .dt <- left_join(.bulk, .sc) %>%
        mutate(sc.z = pmin(pmax(z, -10), 10)) %>%
        na.omit %>%
        arrange(abs(sc.z)) %>% 
        as.data.table

    .dt.show <-
        .dt[(fwer < fwer.cutoff) & bulk.qv < qv.cutoff] %>%
        group_by(sign(sc.z)) %>%
        top_n(10, -log10(pv)) %>%
        ungroup() %>%
        as.data.table

    rr <- cor(.dt$z, .dt$bulk.t, method="spearman")
    r.pv <- cor.test(.dt$z, .dt$bulk.t, method="spearman")$p.value

    .dt.pos <- .dt.show[z > 0]
    .dt.neg <- .dt.show[z < 0]

    .scale.size <-
        scale_size_continuous(range=c(1,2.5), labels = function(x) num.sci(10^(-x)), guide = FALSE)

    plt <-
        ggplot(.dt, aes(x = bulk.t, y = sc.z, size = -log10(pmax(pv, bulk.pv)))) +
        theme_minimal() +
        .scale.size +
        xlab("bulk RNA-seq DEG effect") + ylab("scRNA-seq DEG effect") +
        scale_x_continuous(limits = c(-7, 7)) +
        geom_point(stroke=.2, pch=21, colour="gray40") +
        geom_vline(xintercept = 0, size = .1) +
        geom_hline(yintercept = 0, size = .1) +
        geom_smooth(method="lm", se=FALSE, colour="red", lty=1, size=.5)

    .aes <- aes(label=gene %&% "\n(" %&% num.sci(bulk.pv) %&% ")")

    if(nrow(.dt.pos) > 0) {
        plt <- plt +
            geom_point(data=.dt.pos, pch=21, fill="#ef8a62") +
            ggrepel::geom_label_repel(data=.dt.pos, .aes,
                                     size = 3, vjust=1, hjust=1,
                                     segment.size = .1,
                                     segment.color = "#ef8a62",
                                     box.padding = .1,
                                     label.padding = .1,
                                     nudge_x = -1, nudge_y = -1,
                                     max.overlaps = 10)
    }

    if(nrow(.dt.neg) > 0) {
        plt <- plt +
            geom_point(data=.dt.neg, pch=21, fill="#67a9cf") +
            ggrepel::geom_label_repel(data=.dt.neg, .aes,
                                     size = 3, vjust=0, hjust=0,
                                     segment.size = .1,
                                     segment.color = "#67a9cf",
                                     box.padding = .1,
                                     label.padding = .1,
                                     nudge_x = 1, nudge_y = 1,
                                     max.overlaps = 0)
    }

    plt <- plt +
        ggtitle("Cor=" %&%
                num.round(rr) %&%
                ", p=" %&% num.sci(r.pv) %&%
                ", N=" %&% num.int(nrow(.dt)))

    list(plt=plt, cor = rr, cor.pv = r.pv)
}

show.each.gene <- function(g, .data, .stat, .cutoff = 1e-4) {

    .dt.g <- .data[gene == g]
    .stat.dt <- .stat[gene == g]

    .dir <- .stat.dt[which.max(abs(.stat.dt$ADE)), .(ADE)] %>%
        unlist
    
    if(.dir > 0) {

        .dt <- .dt.g %>%
            filter(mu > log(.cutoff)) %>% 
            left_join(.stat.dt) %>% 
            group_by(disease, celltype) %>% 
            arrange(cfa) %>%
            mutate(j = 1:n()) %>%
            ungroup() %>% 
            mutate(x = disease %&% "_" %&% j)

    } else {

        .dt <- .dt.g %>%
            filter(mu > log(.cutoff)) %>% 
            left_join(.stat.dt) %>% 
            group_by(disease, celltype) %>% 
            arrange(desc(cfa)) %>%
            mutate(j = 1:n()) %>%
            ungroup() %>% 
            mutate(x = disease %&% "_" %&% j)

    }

    .dt <- .dt %>%
        mutate(.pv = if_else(is.na(pv), "1", num.sci(pv))) %>% 
        mutate(.facet = celltype %&% "\n(" %&% .pv %&% ")") %>%
        mutate(ADD.z = ADD/ADD.se, ADC.z = ADC/ADC.se) %>%
        as.data.table

    .stat.dt <- .dt[, .(celltype,
                        ADE, ADE.se,
                        ADD, ADD.se,
                        ADC, ADC.se,
                        pv, .facet)] %>%
        unique

    .aes <- aes(x = x,
                y = cfa,
                ymin = cfa - 2 * cfa.sd,
                ymax = cfa + 2 * cfa.sd,
                fill = disease)

    .scale.y <- scale_y_continuous(labels = function(x) num.round(exp(x)))

    p1 <-
        .gg.plot(.dt, .aes) +
        theme(legend.position = "top") +
        facet_wrap(~.facet, nrow = 1, scales="free") +
        .scale.y +
        theme(axis.ticks.x = element_blank()) +
        theme(axis.text.x = element_blank()) +
        theme(axis.title.x = element_blank()) +
        geom_hline(yintercept = 0, size = 1, colour = "gray") +
        geom_hline(data = .stat.dt, aes(yintercept = ADE), size = 1, colour = "#F8766D") +
        geom_linerange(size = .2) +
        geom_point(aes(size = exp(mu)), pch = 21, stroke = .2) +
        scale_size_continuous("mean", range=c(0, 2.5)) +
        scale_fill_manual(values = c("gray90","#F8766D")) +
        xlab("") + ylab("disase effect") + ggtitle(g)

    .temp.1 <- .stat.dt %>%
        dplyr::select(celltype, pv, ADE, ADD, ADC) %>%
        mutate(ADC = - ADC) %>%                  # flip the effect
        melt(id.vars=c("celltype", "pv"), value.name = "mu")

    .temp.2 <- .stat.dt %>%
        dplyr::select(celltype, ADE.se, ADD.se, ADC.se) %>%
        melt(id.vars="celltype", value.name = "se") %>%
        mutate(variable = str_remove(variable, ".se"))

    .temp <- left_join(.temp.1, .temp.2)

    p2 <- .gg.plot(.temp, aes(x = variable, y = mu, ymin = mu - 3 * se, ymax = mu + 3 * se)) +
        theme(axis.text.x = element_text(angle=90, hjust = 1, vjust= .5)) +
        facet_wrap(~celltype, nrow = 1, scales="free") +
        .scale.y + ylab("disease effect") + xlab("") +
        geom_hline(yintercept = 0, lty = 2, size = .5, colour = "black") +
        geom_linerange(size = .2) +
        geom_point(aes(fill = variable, stroke = .2, size = -log10(pv)), pch = 21) +
        scale_fill_brewer(palette = "Greens", guide=FALSE) +
        scale_size_continuous("p-value", range=c(.5, 2.5), breaks=seq(0, 10, 2), labels = function(x) 10^(-x))

    plt <- (p1 / p2) + patchwork::plot_layout(heights = c(4,2))
    return(plt)
}

count.deg <- function(.dt, fwer.cutoff = .05) {
    .dt[fwer < fwer.cutoff, .(n = .N),
        by = .(celltype,
               direction = if_else(z > 0, "up", "down"))
        ] %>%
        mutate(direction = factor(direction, c("up", "down"))) %>%
        group_by(celltype) %>%
        arrange(desc(direction)) %>%
        mutate(nc = cumsum(n)) %>% 
        ungroup
}
```

# Result: MS vs. HC for mTreg T-cells

```{r read_annot}
.col <- c("tag", "celltype", "prob", "ln.prob")
round1.dt <- .fread("result/step2/assignment/round1.annot.gz", col.names=.col) %>%
    mutate(j = 1:n()) %>%
    as.data.table

.col <- c("tag", "celltype.th", "prob.th", "ln.prob.th")
round2.dt <- .fread("result/step3/assignment/Memory.annot.gz", col.names=.col)

annot.dt <- left_join(round1.dt, round2.dt) %>%
    filter(!is.na(celltype.th)) %>%
    filter(celltype == "mTreg") %>% 
    mutate(celltype = str_replace(celltype.th, "TH", "mTreg")) %>% 
    as.data.table

annot.dt[, c("barcode","batch") := tstrsplit(tag, split="_")]
```

## DEG MS vs. HC in pseudo-bulk analysis

```{r select_deg_cells}
.mkdir("result/step5/deg")
.mkdir("result/step5/temp/")

.hash <- fread("data/hashtag.tsv.gz", header=TRUE)
.cols <- .fread("result/step1/matrix_final.cols.gz", col.names="tag")
.cols[, c("barcode","batch") := tstrsplit(tag, "_")]
.cols[, batch := as.integer(batch)]
.cols <- left_join(.cols, .hash) %>% distinct

.temp <-
    annot.dt %>%
    mutate(batch = as.integer(batch)) %>%
    left_join(.cols) %>%
    filter(!is.na(disease), !is.na(celltype)) %>%
    filter(celltype != "CD8") %>%
    dplyr::select(celltype, tag) %>%
    distinct()

.mkdir("result/step5/temp")
.fwrite(unique(.temp[, .(tag)]), "result/step5/temp/ms_vs_hc.select.gz")
```


```{sh select_annotated_disease_cells}
[ -d result/step5/deg ] || mkdir -p result/step5/deg/
[ -d result/step5/aggregate/ ] || mkdir -p result/step5/aggregate/

if ! [ -f result/step5/deg/ms_vs_hc.mtx.gz ] ; then
	
	[ -f result/step5/temp/ms_vs_hc.mtx.gz ] || \
		mmutil_select_col \
			result/step1/matrix_final.mtx.gz \
			result/step1/matrix_final.cols.gz \
			result/step5/temp/ms_vs_hc.select.gz \
			result/step5/temp/ms_vs_hc \
		2> /dev/null

	mmutil_filter_row \
		--mtx result/step5/temp/ms_vs_hc.mtx.gz \
		--row result/step1/matrix_final.rows.gz \
		--col result/step5/temp/ms_vs_hc.cols.gz \
		--score NNZ \
		--cutoff 100 \
		--col_cutoff 500 \
		--out result/step5/deg/ms_vs_hc \
		2> /dev/null

	rm -f result/step5/temp/ms_vs_hc.*
fi
```

```{r create_cfa_input, include = FALSE}
.cols <- .fread("result/step5/deg/ms_vs_hc.cols.gz", col.names="tag")
.hash <- fread("data/hashtag.tsv.gz", header=TRUE)

.input <- .cols %>%
    mutate(.tag = tag) %>%
    separate(tag, c("barcode", "batch"), sep="[_]") %>%
    left_join(annot.dt) %>%
    mutate(batch = as.integer(batch)) %>%
    left_join(.hash, by = c("barcode", "batch")) %>%
    mutate(ind = batch %&% "_" %&% disease) %>%
    as.data.table

.fwrite(.input[,.(.tag, celltype)], "result/step5/deg/ms_vs_hc.annot.gz")
.fwrite(.input[,.(batch)], "result/step5/deg/ms_vs_hc.batch.gz")
.fwrite(.input[,.(ind)], "result/step5/deg/ms_vs_hc.ind.gz")
.fwrite(.input[,.(disease)], "result/step5/deg/ms_vs_hc.trt.gz")
.fwrite(distinct(.input[,.(celltype)]), "result/step5/deg/ms_vs_hc.lab.gz")
```

```{sh run_cocoa_ms_vs_hc}
mkdir -p result/step5/deg

[ -f result/step5/deg/bbknn_ms_vs_hc.factors.gz ] || \
	mmutil_bbknn \
		--mtx result/step5/deg/ms_vs_hc.mtx.gz \
		--col result/step5/deg/ms_vs_hc.cols.gz \
		--batch result/step5/deg/ms_vs_hc.batch.gz \
		--knn 100 --rank 10 --log_scale \
		--out result/step5/deg/bbknn_ms_vs_hc \
		2> /dev/null

[ -f result/step5/deg/ms_vs_hc.resid_mu.gz ] || \
	mmutil_cfa_col \
		--mtx result/step5/deg/ms_vs_hc.mtx.gz \
		--col result/step5/deg/ms_vs_hc.cols.gz \
		--annot result/step5/deg/ms_vs_hc.annot.gz \
		--trt result/step5/deg/ms_vs_hc.trt.gz \
		--ind result/step5/deg/ms_vs_hc.ind.gz \
		--lab result/step5/deg/ms_vs_hc.lab.gz \
		--knn 100 --rank 10 --nthreads 10 \
		--gamma_a0 1 --gamma_b0 1 --impute_knn \
		--verbose --log_scale \
		--out result/step5/deg/ms_vs_hc \
		2> /dev/null
```


```{r read_ms_vs_hc_results, results="asis", echo = FALSE}
.mkdir("Tab")
.file <- "Tab/DEG_mTreg_MS_vs_HC.txt.gz"
if(!file.exists(.file)) {
    deg.dt <-
        read.deg.result("result/step5/deg/ms_vs_hc") %>%
        summarize.deg(expr.cutoff = 1e-2)

    .fwrite.safe(deg.dt, .file)
}
deg.dt <- fread(.file)

cat("[mTreg DEG_MS_vs_HC download]("%&% .file %&% ")\n\n")
```


## Pseudo-bulk data

```{r read_pseudo_bulk_data, results = "asis", echo = FALSE}
.mkdir("Tab")
.file <- "Tab/MS_vs_HC_mTreg_pseudobulk_data.txt.gz"
if(!file.exists(.file)) {
    pb.data <- read.deg.result("result/step5/deg/ms_vs_hc")
    .fwrite.safe(pb.data, .file)
}

pb.data <- fread(.file, header=TRUE)
cat("[DEG_MS_vs_HC pseudo-bulk data download]("%&% .file %&% ")\n\n")
```

* Error bars in the top panels: 2 times standard error for each sample

* Error bars in the bottom panels: 3 times standard error for each type of effect

* ADC: Average Disease effect in the Control samples

* ADD: Average Disease effect in the Disease samples

* ADE: Average Disease Effect across all the samples

### Found `r num.int(nrow(unique(deg.dt[fwer < .05, .(gene)])))` unique genes strongly perturbed by MS with FWER 5%

* Up-regulated: `r num.int(nrow(unique(deg.dt[fwer < .05 & z > 0, .(gene)])))`

* Down-regulated:  `r num.int(nrow(unique(deg.dt[fwer < .05 & z < 0, .(gene)])))`

* Total pairs of genes and cell types: `r num.int(nrow(deg.dt))`

```{r Fig_DEG_count, fig.width=2, fig.height=2, echo = FALSE, results="asis"}
.df <- count.deg(deg.dt)

plt <-
    .gg.plot(.df, aes(x = celltype, y = n, group = direction, fill = direction)) +
    theme(legend.position = c(1,1)) +
    theme(legend.justification = c(1,1)) +
    xlab("cell type") + ylab("# DEG") +
    geom_bar(stat="identity") +
    geom_text(aes(y=nc, label=n)) +
    scale_fill_manual(values = c("#ef8a62","#67a9cf"))

print(plt)
.file <- fig.dir %&% "/Fig_DEG_count.pdf"
.gg.save(filename = .file, plot = plt, width=2, height=2)
```

```{r}
.multi <-
    deg.dt[fwer < 1e-2, .(n = .N), by = .(gene)] %>%
    dplyr::filter(n > 1) %>%
    .select(gene)
```

## Cell-type-specific gene ontology enrichment, removing `r num.int(length(.multi))` genes

```{r test_go}
run.goseq <- function(.dt, .ct, fwer.cutoff, strict.bg = FALSE) {

    if(strict.bg){
        .temp <- .dt[celltype == .ct]
    } else {
        .temp <- .dt
    }

    .genes <- .temp[, .(ensembl_gene_id)] %>%
        unique %>%
        unlist

    .de.genes <- .temp[celltype == .ct &
                       fwer < fwer.cutoff,
                       .(ensembl_gene_id)] %>%
        unique %>%
        unlist

    gene.vector <- as.integer(.genes %in% .de.genes)
    names(gene.vector) <- .genes
    pwf <- goseq::nullp(gene.vector, "hg19", "ensGene", plot.fit=FALSE)
    go.dt <- goseq::goseq(pwf, "hg19", "ensGene", use_genes_without_cat = TRUE) %>%
        as.data.table
}

plot.go.heatmap <- function(go.ct.dt, .celltype.order, fdr.cutoff = .1, .ntop = NULL) {

    if(is.null(.ntop)) {
        go.show <- go.ct.dt[qv < fdr.cutoff & numDEInCat > 2,
                            .(term)] %>% unique
    } else {
        go.show <- go.ct.dt[qv < fdr.cutoff & numDEInCat > 2] %>%
            group_by(celltype) %>%
            top_n(.ntop, numDEInCat/numInCat) %>%
            ungroup %>%
            dplyr::select(term) %>%
            unique
    }

    .dt <-
        merge(go.show, go.ct.dt) %>%
        mutate(row = celltype, col = term, weight = -log10(pv)) %>%
        col.order(.ro = .celltype.order, ret.tab = TRUE) %>%
        as.data.table

    .aes <- aes(y = col, x = row,
                size = -log10(pmax(pv, 1e-6)),
                fill = log10(numDEInCat))

    .scale.fill <-
        scale_fill_distiller("#genes",
                             palette = "RdPu",
                             direction = 1,
                             labels=function(x) round(10^(x)))

    .scale.size <-
        scale_size_continuous("p-value",
                              breaks = seq(0,10,2),
                              range=c(0, 3),
                              labels=function(x) num.sci(10^(-x)))

    plt <-
        .gg.plot(.dt, .aes) +
        theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
        theme(legend.key.width = unit(.2, "lines")) +
        theme(legend.key.height = unit(.5, "lines")) +
        geom_point(pch = 22, stroke=.1) +
        .scale.fill +
        xlab("cell type") + ylab("gene ontology") +
        .scale.size

    .nrow <- length(unique(.dt$term))
    .ncol <- length(unique(.dt$celltype))
    .lab <- max(sapply(.dt$term, nchar))

    list(plt = plt,
         w = (.lab * .05 + .ncol * .075 + 1),
         h = ceiling(.1 * .nrow + .25))
}

save.heatmap <- function(.obj, .file) {
    .mkdir(dirname(.file))
    .gg.save(filename=.file, cat.link=FALSE, .obj$plt, width = .obj$w, height = .obj$h)
}

.dt <-
    deg.dt %>%
    dplyr::filter(!(gene %in% .multi)) %>% 
    dplyr::rename(hgnc_symbol = gene) %>%
    left_join(fread("data/gene.info.gz")) %>%
    na.omit

.ct.order <- c("mTreg17","mTreg1/17","mTreg1","nTreg2")

go.result.dt <- 
    lapply(.ct.order,
           function(.ct)
               run.goseq(.dt, .ct, fwer.cutoff = 0.05) %>%
               mutate(celltype = .ct)
           ) %>%
    do.call(what=rbind) %>%
    mutate(pv = over_represented_pvalue) %>% 
    mutate(qv = p.adjust(pv, "fdr")) %>% 
    as.data.table
```

```{r Fig_GO_, results = "asis", echo = FALSE}
plt <- 
    go.result.dt[ontology == "BP"] %>%
    plot.go.heatmap(.ct.order, fdr.cutoff = 0.05, .ntop=15)

.file <- fig.dir %&% "/Fig_GO_BP.pdf"
save.heatmap(plt, .file)
cat("[GO BP](" %&% .file %&% ") ")

plt <- 
    go.result.dt[ontology == "CC"] %>%
    plot.go.heatmap(.ct.order, fdr.cutoff = 0.05, .ntop=15)

.file <- fig.dir %&% "/Fig_GO_CC.pdf"
save.heatmap(plt, .file)
cat("[GO CC](" %&% .file %&% ") ")
```

#### Show genes with FWER $< 0.05$ and some other genes

```{r Fig_DEG_example, results="asis", echo = FALSE}
.genes.ms <-
    deg.dt[fwer < 0.05 &
           sign(ADD) == sign(z) &
           sign(ADC) == sign(z),
           .(gene)] %>%
    unique %>% unlist

.genes.ms <- c(.genes.ms, "SGK1") %>%
    unique

for(g in .genes.ms){
    .file <- fig.dir %&% "/Fig_DEG_example_" %&% g %&% ".pdf"
    if(!file.exists(.file)) {
        plt <- show.each.gene(g,
                              pb.data[celltype != "TFH"],
                              deg.dt[celltype != "TFH"])
        .gg.save(filename = .file, cat.link = FALSE, plot = plt, width=5, height=3)
    }
    cat("[" %&% g %&% "](" %&% .file %&% ") ")
}
cat("\n\n\n")
```
